// classpresso/dist/cli.cjs
'use strict';

var promises = require('fs/promises');
var glob = require('glob');
var path3 = require('path');
var commander = require('commander');
var chalk2 = require('chalk');
var module$1 = require('module');
var url = require('url');
var os = require('os');
var postcss = require('postcss');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var path3__default = /*#__PURE__*/_interopDefault(path3);
var chalk2__default = /*#__PURE__*/_interopDefault(chalk2);
var os__default = /*#__PURE__*/_interopDefault(os);
var postcss__default = /*#__PURE__*/_interopDefault(postcss);

var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/utils/files.ts
var files_exports = {};
__export(files_exports, {
  DEFAULT_PATTERNS: () => DEFAULT_PATTERNS,
  backupFile: () => backupFile,
  filterExcludedFiles: () => filterExcludedFiles,
  findFiles: () => findFiles,
  getFileSize: () => getFileSize,
  getFileStats: () => getFileStats,
  isCSSFile: () => isCSSFile,
  isHTMLFile: () => isHTMLFile,
  isJSFile: () => isJSFile,
  isRSCFile: () => isRSCFile,
  readFileContent: () => readFileContent,
  shouldExcludeFile: () => shouldExcludeFile,
  writeFileContent: () => writeFileContent
});
async function findFiles(buildDir, patterns) {
  const globPatterns = [...new Set(patterns.map((p) => path3__default.default.join(buildDir, p)))];
  const allFiles = await glob.glob(globPatterns, {
    nodir: true,
    windowsPathsNoEscape: true
  });
  return [...new Set(allFiles)];
}
async function readFileContent(filePath) {
  return promises.readFile(filePath, "utf-8");
}
async function writeFileContent(filePath, content) {
  await promises.writeFile(filePath, content, "utf-8");
}
async function getFileSize(filePath) {
  const stats = await promises.stat(filePath);
  return stats.size;
}
async function backupFile(filePath) {
  const backupPath = `${filePath}.classpresso-backup`;
  await promises.copyFile(filePath, backupPath);
  return backupPath;
}
async function getFileStats(filePaths) {
  const stats = [];
  for (const filePath of filePaths) {
    try {
      const size = await getFileSize(filePath);
      stats.push({
        path: filePath,
        originalSize: size,
        modified: false
      });
    } catch {
    }
  }
  return stats;
}
function isJSFile(filePath) {
  return /\.(js|mjs|cjs)$/.test(filePath);
}
function isCSSFile(filePath) {
  return /\.css$/.test(filePath);
}
function isHTMLFile(filePath) {
  return /\.(html|htm)$/.test(filePath);
}
function isRSCFile(filePath) {
  return /\.rsc$/.test(filePath);
}
function shouldExcludeFile(filePath, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return false;
  }
  const normalizedPath = filePath.replace(/\\/g, "/");
  for (const pattern of excludePatterns) {
    const regexPattern = pattern.replace(/\\/g, "/").replace(/\*\*/g, "{{DOUBLESTAR}}").replace(/\*/g, "[^/]*").replace(/\?/g, ".").replace(/{{DOUBLESTAR}}/g, ".*");
    const regex = new RegExp(regexPattern);
    if (regex.test(normalizedPath)) {
      return true;
    }
  }
  return false;
}
function filterExcludedFiles(files, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return files;
  }
  return files.filter((filePath) => !shouldExcludeFile(filePath, excludePatterns));
}
var DEFAULT_PATTERNS;
var init_files = __esm({
  "src/utils/files.ts"() {
    DEFAULT_PATTERNS = [
      // === Next.js (.next directory) ===
      // Client-side chunks
      "static/chunks/**/*.js",
      "static/css/**/*.css",
      // Server components (Next.js App Router)
      "server/app/**/*.js",
      "server/app/**/*.html",
      "server/app/**/*.rsc",
      // Server chunks (contains Client Component code used for hydration)
      "server/chunks/**/*.js",
      // Standalone build (Next.js output: 'standalone')
      "standalone/.next/static/chunks/**/*.js",
      "standalone/.next/static/css/**/*.css",
      "standalone/.next/server/app/**/*.js",
      "standalone/.next/server/app/**/*.html",
      "standalone/.next/server/app/**/*.rsc",
      "standalone/.next/server/chunks/**/*.js",
      // === Nuxt (.output directory) ===
      "public/**/*.html",
      "public/**/*.js",
      "public/**/*.css",
      "public/_nuxt/**/*.js",
      "public/_nuxt/**/*.css",
      "server/**/*.mjs",
      "server/chunks/**/*.mjs",
      // === SvelteKit (build directory) ===
      // Static adapter output
      "**/*.html",
      "_app/**/*.js",
      "_app/**/*.css",
      // Node adapter output
      "client/**/*.js",
      "client/**/*.css",
      "server/**/*.js",
      // === Astro (dist directory) ===
      "_astro/**/*.js",
      "_astro/**/*.css",
      // SSR/Hybrid build - Client assets
      "client/_astro/**/*.js",
      "client/_astro/**/*.css",
      // === Remix (build directory) ===
      "client/**/*.js",
      "client/**/*.css",
      "server/**/*.js",
      // === Solid Start (.output or dist directory) ===
      "public/**/*.html",
      "public/**/*.js",
      "public/**/*.css",
      "_build/**/*.js",
      "_build/**/*.css",
      // === Qwik (dist directory) ===
      "build/**/*.js",
      "build/**/*.css",
      "q-*.js",
      // === Angular (dist/[project] directory) ===
      // Angular 17+ uses browser subdirectory
      "browser/**/*.js",
      "browser/**/*.css",
      "browser/**/*.html",
      // Legacy Angular output
      "*.js",
      "*.css",
      "*.html",
      // === Gatsby (public directory) ===
      "page-data/**/*.json",
      "*.js",
      "*.css",
      "*.html",
      // === Vite/Vue/React generic (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "index.html",
      // === Ember (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "*.html",
      // === VitePress (.vitepress/dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === Docusaurus (build directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === Eleventy/11ty (_site directory) ===
      "**/*.html",
      "**/*.js",
      "**/*.css",
      // === Hugo (public directory) ===
      "**/*.html",
      "**/*.js",
      "**/*.css",
      // === Parcel (dist directory) ===
      "*.html",
      "*.js",
      "*.css",
      // === Preact (build directory) ===
      "bundle.js",
      "bundle.*.js",
      "*.css",
      "*.html",
      // === Gridsome (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === RedwoodJS (web/dist directory) ===
      "static/**/*.js",
      "static/**/*.css",
      "*.html",
      // === Webpack generic (dist directory) ===
      "*.js",
      "*.css",
      "*.html",
      "js/**/*.js",
      "css/**/*.css"
    ];
  }
});

// src/core/scanner.ts
init_files();

// src/utils/regex.ts
function escapeRegex(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
var CLASS_PATTERNS = {
  // JSX: className="..."
  jsxDouble: /className\s*=\s*"([^"]+)"/g,
  // JSX: className='...'
  jsxSingle: /className\s*=\s*'([^']+)'/g,
  // React createElement: className:"..."
  createElementDouble: /className\s*:\s*"([^"]+)"/g,
  // React createElement: className:'...'
  createElementSingle: /className\s*:\s*'([^']+)'/g,
  // Minified: "className","..."
  minifiedComma: /"className"\s*,\s*"([^"]+)"/g,
  // HTML: class="..."
  htmlDouble: /\bclass\s*=\s*"([^"]+)"/g,
  // HTML: class='...'
  htmlSingle: /\bclass\s*=\s*'([^']+)'/g,
  // HTML entity encoded: class=&quot;...&quot;
  htmlEntityDouble: /\bclass=&quot;([^&]+)&quot;/g,
  // HTML entity encoded: class=&#34;...&#34; (numeric entity)
  htmlNumericDouble: /\bclass=&#34;([^&]+)&#34;/g,
  // RSC payload: \"className\":\"...\" (escaped JSON in Next.js RSC payloads)
  rscPayload: /\\"className\\"\s*:\s*\\"([^"\\]+)\\"/g,
  // Vue/Quasar createElement: class:"..." (Vue uses 'class' not 'className')
  vueCreateElementDouble: /\bclass\s*:\s*"([^"]+)"/g,
  // Vue/Quasar createElement: class:'...'
  vueCreateElementSingle: /\bclass\s*:\s*'([^']+)'/g
};
var ALL_CLASS_PATTERNS = [
  CLASS_PATTERNS.jsxDouble,
  CLASS_PATTERNS.jsxSingle,
  CLASS_PATTERNS.createElementDouble,
  CLASS_PATTERNS.createElementSingle,
  CLASS_PATTERNS.minifiedComma,
  CLASS_PATTERNS.htmlDouble,
  CLASS_PATTERNS.htmlSingle,
  CLASS_PATTERNS.htmlEntityDouble,
  CLASS_PATTERNS.htmlNumericDouble,
  CLASS_PATTERNS.rscPayload,
  CLASS_PATTERNS.vueCreateElementDouble,
  CLASS_PATTERNS.vueCreateElementSingle
];
function isDynamicClassString(classString) {
  if (classString.includes("${")) return true;
  if (/\s*\?\s*/.test(classString) && classString.includes(":")) return true;
  if (/\w+\s*\(/.test(classString)) return true;
  if (/^[a-z][a-zA-Z0-9]*$/.test(classString) && !classString.includes("-")) return true;
  return false;
}
var DYNAMIC_BASE_PATTERNS = [
  // className:`base classes ${...}`
  /className\s*:\s*`([^`$]+)\s*\$\{/g,
  // className=`base classes ${...}` (JSX)
  /className\s*=\s*`([^`$]+)\s*\$\{/g,
  // class:`base classes ${...}` (rare but possible)
  /\bclass\s*:\s*`([^`$]+)\s*\$\{/g
];
var CONCAT_BASE_PATTERNS = [
  // className:"".concat("base classes",...) - the base before concat
  /className\s*:\s*""\s*\.concat\s*\(\s*"([^"]+)"/g,
  // className:''.concat('base classes',...)
  /className\s*:\s*''\s*\.concat\s*\(\s*'([^']+)'/g
];
function extractConcatBaseStrings(content) {
  const results = [];
  for (const pattern of CONCAT_BASE_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const baseString = match[1]?.trim();
      if (baseString && baseString.length > 0) {
        results.push(baseString);
      }
    }
  }
  return results;
}
function extractDynamicBaseStrings(content) {
  const results = [];
  for (const pattern of DYNAMIC_BASE_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const baseString = match[1]?.trim();
      if (baseString && baseString.length > 0) {
        results.push(baseString);
      }
    }
  }
  return results;
}

// src/core/scanner.ts
function getSourceFileType(filePath) {
  if (isHTMLFile(filePath)) return "html";
  if (isRSCFile(filePath)) return "rsc";
  return "js";
}
function isProperSubset(classesA, classesB) {
  if (classesA.length >= classesB.length) return false;
  const setB = new Set(classesB);
  return classesA.every((cls) => setB.has(cls));
}
function detectMergeablePatterns(occurrences) {
  const mergeablePatterns = /* @__PURE__ */ new Set();
  const jsPatterns = [];
  const htmlPatterns = [];
  for (const [, occurrence] of occurrences) {
    if (occurrence.sourceTypes.has("js")) {
      jsPatterns.push(occurrence);
    }
    if (occurrence.sourceTypes.has("html") || occurrence.sourceTypes.has("rsc")) {
      htmlPatterns.push(occurrence);
    }
  }
  for (const jsPattern of jsPatterns) {
    for (const htmlPattern of htmlPatterns) {
      if (isProperSubset(jsPattern.classes, htmlPattern.classes)) {
        mergeablePatterns.add(jsPattern.normalizedKey);
        break;
      }
    }
  }
  return mergeablePatterns;
}
function containsDynamicPrefix(classes, dynamicPrefixes) {
  for (const cls of classes) {
    for (const prefix of dynamicPrefixes) {
      if (prefix.endsWith("-")) {
        if (cls.startsWith(prefix)) {
          return true;
        }
      } else {
        if (cls === prefix || cls.startsWith(prefix + "-")) {
          return true;
        }
      }
    }
  }
  return false;
}
function isHydrationSafe(occurrence) {
  const hasServerContext = occurrence.sourceTypes.has("html") || occurrence.sourceTypes.has("rsc");
  const hasClientContext = occurrence.sourceTypes.has("js");
  return hasServerContext && hasClientContext;
}
function shouldExcludeClass(className, exclude) {
  if (exclude.prefixes) {
    for (const prefix of exclude.prefixes) {
      if (className.startsWith(prefix)) return true;
    }
  }
  if (exclude.suffixes) {
    for (const suffix of exclude.suffixes) {
      if (className.endsWith(suffix)) return true;
    }
  }
  if (exclude.classes) {
    if (exclude.classes.includes(className)) return true;
  }
  if (exclude.patterns) {
    for (const pattern of exclude.patterns) {
      if (pattern.test(className)) return true;
    }
  }
  return false;
}
function normalizeClassString(classString, exclude) {
  const allClasses = classString.split(/\s+/).filter(Boolean);
  const includedClasses = [];
  const excludedClasses = [];
  for (const cls of allClasses) {
    if (shouldExcludeClass(cls, exclude)) {
      excludedClasses.push(cls);
    } else {
      includedClasses.push(cls);
    }
  }
  const sorted = [...includedClasses].sort();
  return {
    normalized: sorted.join(" "),
    classes: includedClasses,
    excludedClasses
  };
}
function extractClassStrings(content, filePath) {
  const results = [];
  const seen = /* @__PURE__ */ new Set();
  for (const pattern of ALL_CLASS_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const classString = match[1];
      if (!classString || !classString.trim()) continue;
      if (isDynamicClassString(classString)) continue;
      const key = `${filePath}:${match.index}:${classString}`;
      if (seen.has(key)) continue;
      seen.add(key);
      const beforeMatch = content.substring(0, match.index);
      const line = (beforeMatch.match(/\n/g) || []).length + 1;
      results.push({
        classString: classString.trim(),
        location: {
          filePath,
          line
        }
      });
    }
  }
  return results;
}
async function scanBuildOutput(config) {
  const occurrences = /* @__PURE__ */ new Map();
  const dynamicBasePatterns = /* @__PURE__ */ new Map();
  const fileStats = [];
  const errors = [];
  const patterns = config.include.length > 0 ? config.include : DEFAULT_PATTERNS;
  const allFiles = await findFiles(config.buildDir, patterns);
  const files = filterExcludedFiles(allFiles, config.exclude.files || []);
  if (config.verbose) {
    console.log(`Found ${allFiles.length} files, scanning ${files.length} (${allFiles.length - files.length} excluded)`);
  }
  for (const filePath of files) {
    try {
      if (!isJSFile(filePath) && !isHTMLFile(filePath) && !isRSCFile(filePath)) {
        continue;
      }
      const content = await readFileContent(filePath);
      const size = await getFileSize(filePath);
      fileStats.push({
        path: filePath,
        originalSize: size,
        modified: false
      });
      if (isJSFile(filePath)) {
        const dynamicBases = extractDynamicBaseStrings(content);
        const concatBases = extractConcatBaseStrings(content);
        const allDynamicBases = [...dynamicBases, ...concatBases];
        for (const baseString of allDynamicBases) {
          const { normalized, classes } = normalizeClassString(baseString, config.exclude);
          if (classes.length < config.minClasses) continue;
          const location = { filePath };
          if (dynamicBasePatterns.has(normalized)) {
            dynamicBasePatterns.get(normalized).locations.push(location);
          } else {
            dynamicBasePatterns.set(normalized, {
              baseClasses: classes,
              normalizedKey: normalized,
              locations: [location]
            });
          }
        }
      }
      const classStrings = extractClassStrings(content, filePath);
      const sourceType = getSourceFileType(filePath);
      for (const { classString, location } of classStrings) {
        const { normalized, classes, excludedClasses } = normalizeClassString(
          classString,
          config.exclude
        );
        if (classes.length === 0) continue;
        if (classes.length < config.minClasses) continue;
        if (config.excludeDynamicPatterns && containsDynamicPrefix(classes, config.dynamicPrefixes)) {
          continue;
        }
        if (occurrences.has(normalized)) {
          const existing = occurrences.get(normalized);
          existing.count++;
          existing.locations.push(location);
          existing.sourceTypes.add(sourceType);
        } else {
          occurrences.set(normalized, {
            classString,
            normalizedKey: normalized,
            count: 1,
            locations: [location],
            classes,
            excludedClasses,
            sourceTypes: /* @__PURE__ */ new Set([sourceType])
          });
        }
      }
    } catch (err) {
      errors.push(`Error scanning ${filePath}: ${err}`);
    }
  }
  const mergeablePatterns = detectMergeablePatterns(occurrences);
  if (config.verbose && mergeablePatterns.size > 0) {
    console.log(`Found ${mergeablePatterns.size} patterns that appear to be merged className props`);
  }
  return {
    occurrences,
    files: fileStats,
    errors,
    dynamicBasePatterns,
    mergeablePatterns
  };
}

// src/utils/hash.ts
var ALPHABET = "abcdefghijklmnopqrstuvwxyz0123456789";
var BASE = ALPHABET.length;
function generateSequentialName(index, prefix = "cp-") {
  let result = "";
  let n = index;
  do {
    result = ALPHABET[n % BASE] + result;
    n = Math.floor(n / BASE) - 1;
  } while (n >= 0);
  return `${prefix}${result}`;
}

// src/core/metrics.ts
function calculateMetrics(candidates, originalFiles, transformResult, cssOverhead) {
  const totalBytesSaved = candidates.reduce((sum, c) => sum + c.bytesSaved, 0);
  const netSavings = totalBytesSaved - cssOverhead;
  const originalTotalBytes = originalFiles.reduce((sum, f) => sum + f.originalSize, 0);
  const optimizedTotalBytes = originalTotalBytes - netSavings;
  const estimatedParseTimeSavedMs = Math.max(0, netSavings / 1024 * 0.1);
  const estimatedRenderTimeSavedMs = estimatedParseTimeSavedMs * 0.5;
  return {
    totalFilesScanned: originalFiles.length,
    totalFilesModified: transformResult.filesModified,
    totalClassStringsFound: candidates.reduce((sum, c) => sum + c.frequency, 0),
    uniqueClassPatterns: candidates.length,
    consolidatedPatterns: candidates.filter((c) => c.frequency >= 2).length,
    totalOccurrencesReplaced: candidates.reduce((sum, c) => sum + c.frequency, 0),
    originalTotalBytes,
    optimizedTotalBytes,
    bytesSaved: totalBytesSaved,
    percentageReduction: originalTotalBytes > 0 ? netSavings / originalTotalBytes * 100 : 0,
    originalCSSBytes: 0,
    // Would need to track CSS separately
    consolidatedCSSBytes: cssOverhead,
    netCSSChange: cssOverhead,
    estimatedParseTimeSavedMs,
    estimatedRenderTimeSavedMs,
    topConsolidations: candidates.slice(0, 10).map((c) => ({
      original: c.classString,
      consolidated: c.hashName,
      frequency: c.frequency,
      bytesSaved: c.bytesSaved
    }))
  };
}
function calculatePatternCSSOverhead(classes) {
  const baseOverhead = 15;
  const perClassOverhead = 20;
  return baseOverhead + classes.length * perClassOverhead;
}
function estimateCSSOverhead(candidates) {
  return candidates.reduce((sum, c) => {
    return sum + calculatePatternCSSOverhead(c.classes);
  }, 0);
}
function formatBytes(bytes) {
  if (bytes === 0) return "0 B";
  const sign = bytes < 0 ? "-" : "";
  const absBytes = Math.abs(bytes);
  if (absBytes < 1024) {
    return `${sign}${absBytes} B`;
  } else if (absBytes < 1024 * 1024) {
    return `${sign}${(absBytes / 1024).toFixed(2)} KB`;
  } else {
    return `${sign}${(absBytes / (1024 * 1024)).toFixed(2)} MB`;
  }
}
function formatPercentage(value) {
  return `${value.toFixed(2)}%`;
}
function formatTime(ms) {
  if (ms < 1) {
    return `${(ms * 1e3).toFixed(2)}\u03BCs`;
  } else if (ms < 1e3) {
    return `${ms.toFixed(2)}ms`;
  } else {
    return `${(ms / 1e3).toFixed(2)}s`;
  }
}

// src/core/pattern-detector.ts
function calculateBytesSaved(original, hashName, frequency, excludedClasses) {
  const originalLength = original.length;
  const excludedLength = excludedClasses.join(" ").length;
  const includedLength = originalLength - excludedLength - (excludedClasses.length > 0 ? 1 : 0);
  const newLength = hashName.length;
  const savingsPerOccurrence = includedLength - newLength;
  return savingsPerOccurrence * frequency;
}
function detectConsolidatablePatterns(occurrences, config, mergeablePatterns) {
  const preliminaryCandidates = [];
  for (const [, occurrence] of occurrences) {
    if (occurrence.count < config.minOccurrences) continue;
    if (occurrence.classes.length < config.minClasses) continue;
    if (config.ssr && !isHydrationSafe(occurrence)) {
      continue;
    }
    if (config.ssr && mergeablePatterns && mergeablePatterns.has(occurrence.normalizedKey)) {
      continue;
    }
    if (config.skipPatternsWithExcludedClasses && occurrence.excludedClasses.length > 0) {
      continue;
    }
    const cssOverhead = calculatePatternCSSOverhead(occurrence.classes);
    preliminaryCandidates.push({ occurrence, cssOverhead });
  }
  preliminaryCandidates.sort((a, b) => b.occurrence.count - a.occurrence.count);
  const filteredCandidates = [];
  for (const { occurrence, cssOverhead } of preliminaryCandidates) {
    const estimatedNameLength = config.hashPrefix.length + 1;
    const estimatedBytesSaved = calculateBytesSaved(
      occurrence.classString,
      "x".repeat(estimatedNameLength),
      // placeholder
      occurrence.count,
      occurrence.excludedClasses
    );
    if (estimatedBytesSaved < config.minBytesSaved) continue;
    if (!config.forceAll) {
      if (estimatedBytesSaved <= cssOverhead) continue;
    }
    filteredCandidates.push({ occurrence, cssOverhead });
  }
  const candidates = [];
  for (let i = 0; i < filteredCandidates.length; i++) {
    const { occurrence } = filteredCandidates[i];
    const hashName = generateSequentialName(i, config.hashPrefix);
    const bytesSaved = calculateBytesSaved(
      occurrence.classString,
      hashName,
      occurrence.count,
      occurrence.excludedClasses
    );
    candidates.push({
      classString: occurrence.classString,
      normalizedKey: occurrence.normalizedKey,
      frequency: occurrence.count,
      bytesSaved,
      classes: occurrence.classes,
      excludedClasses: occurrence.excludedClasses,
      hashName
    });
  }
  return candidates.sort((a, b) => b.bytesSaved - a.bytesSaved);
}
function getPatternSummary(candidates) {
  const totalPatterns = candidates.length;
  const totalOccurrences = candidates.reduce((sum, c) => sum + c.frequency, 0);
  const totalBytesSaved = candidates.reduce((sum, c) => sum + c.bytesSaved, 0);
  const totalClasses = candidates.reduce((sum, c) => sum + c.classes.length, 0);
  return {
    totalPatterns,
    totalOccurrences,
    totalBytesSaved,
    avgFrequency: totalPatterns > 0 ? totalOccurrences / totalPatterns : 0,
    avgClassesPerPattern: totalPatterns > 0 ? totalClasses / totalPatterns : 0
  };
}

// src/config.ts
init_files();
var DEFAULT_DYNAMIC_PREFIXES = [
  "lucide",
  // lucide-react icons
  "heroicon",
  // heroicons
  "icon-",
  // generic icon patterns
  "fa-",
  // Font Awesome
  "fas",
  // Font Awesome solid
  "far",
  // Font Awesome regular
  "fab",
  // Font Awesome brands
  "material-icons",
  // Material Design Icons
  "mdi-",
  // Material Design Icons (alternative)
  "ri-",
  // Remix Icons
  "bi-",
  // Bootstrap Icons
  "tabler-",
  // Tabler Icons
  "phosphor-"
  // Phosphor Icons
];
var DEFAULT_CONFIG = {
  buildDir: ".next",
  minOccurrences: 2,
  minClasses: 2,
  minBytesSaved: 10,
  hashPrefix: "cp-",
  hashLength: 5,
  exclude: {
    prefixes: ["js-", "data-", "hook-", "track-"],
    suffixes: ["-handler", "-trigger"],
    classes: [],
    patterns: [/^qa-/, /^test-/, /^e2e-/],
    files: []
  },
  include: [],
  cssLayer: false,
  dataAttributes: false,
  manifest: true,
  backup: false,
  verbose: false,
  forceAll: false,
  excludeDynamicPatterns: true,
  dynamicPrefixes: DEFAULT_DYNAMIC_PREFIXES,
  skipPatternsWithExcludedClasses: true,
  ssr: false,
  debug: false,
  sendErrorReports: false,
  errorReportUrl: void 0,
  purgeUnusedCSS: false,
  purgeSafelist: [
    /^js-/,
    // JavaScript hooks
    /^data-/,
    // Data attributes used as classes
    /^is-/,
    // State classes
    /^has-/,
    // State classes
    /^active$/,
    /^disabled$/,
    /^hidden$/,
    /^visible$/
  ]
};
async function loadConfig(buildDir) {
  const config = { ...DEFAULT_CONFIG };
  if (buildDir) {
    config.buildDir = buildDir;
  }
  const configPaths = [
    "classpresso.config.js",
    "classpresso.config.mjs",
    "classpresso.config.json"
  ];
  for (const configPath of configPaths) {
    try {
      const fullPath = path3__default.default.resolve(process.cwd(), configPath);
      if (configPath.endsWith(".json")) {
        const content = await readFileContent(fullPath);
        const userConfig = JSON.parse(content);
        return mergeConfig(config, userConfig);
      } else {
        const userConfig = await import(fullPath);
        return mergeConfig(config, userConfig.default || userConfig);
      }
    } catch {
    }
  }
  return config;
}
function mergeConfig(defaults, userConfig) {
  return {
    ...defaults,
    ...userConfig,
    exclude: {
      ...defaults.exclude,
      ...userConfig.exclude,
      // Merge arrays instead of replacing
      prefixes: [
        ...defaults.exclude.prefixes || [],
        ...userConfig.exclude?.prefixes || []
      ],
      suffixes: [
        ...defaults.exclude.suffixes || [],
        ...userConfig.exclude?.suffixes || []
      ],
      classes: [
        ...defaults.exclude.classes || [],
        ...userConfig.exclude?.classes || []
      ],
      patterns: [
        ...defaults.exclude.patterns || [],
        ...userConfig.exclude?.patterns || []
      ],
      files: [
        ...defaults.exclude.files || [],
        ...userConfig.exclude?.files || []
      ]
    },
    // Merge dynamicPrefixes arrays
    dynamicPrefixes: [
      ...defaults.dynamicPrefixes || [],
      ...userConfig.dynamicPrefixes || []
    ],
    // Merge purgeSafelist arrays
    purgeSafelist: [
      ...defaults.purgeSafelist || [],
      ...userConfig.purgeSafelist || []
    ]
  };
}
var require2 = module$1.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('cli.cjs', document.baseURI).href)));
var DEBUG_LOG_FILENAME = "classpresso-debug.log";
function getVersion() {
  try {
    const __filename = url.fileURLToPath((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('cli.cjs', document.baseURI).href)));
    const __dirname = path3__default.default.dirname(__filename);
    const packageJsonPath = path3__default.default.resolve(__dirname, "../package.json");
    const packageJson = require2(packageJsonPath);
    return packageJson.version || "unknown";
  } catch {
    return "unknown";
  }
}
function createLogger(options) {
  const { debug, verbose, buildDir } = options;
  const logPath = debug ? path3__default.default.join(buildDir, DEBUG_LOG_FILENAME) : null;
  let initialized = false;
  async function writeToLog(content) {
    if (!logPath) return;
    try {
      if (!initialized) {
        const header = `=== CLASSPRESSO DEBUG LOG ===
Generated: ${(/* @__PURE__ */ new Date()).toISOString()}

`;
        await promises.writeFile(logPath, header, "utf-8");
        initialized = true;
      }
      await promises.appendFile(logPath, content + "\n", "utf-8");
    } catch (err) {
      console.error(`Warning: Failed to write to debug log: ${err}`);
    }
  }
  function timestamp() {
    return `[${(/* @__PURE__ */ new Date()).toISOString()}]`;
  }
  return {
    info(message) {
      console.log(message);
    },
    warn(message) {
      console.warn(message);
    },
    error(message) {
      console.error(message);
    },
    verbose(message) {
      if (verbose) {
        console.log(message);
      }
    },
    async logSystemInfo() {
      if (!debug) return;
      const systemInfo = [
        "=== SYSTEM INFO ===",
        `Classpresso Version: ${getVersion()}`,
        `Node Version: ${process.version}`,
        `OS: ${process.platform}`,
        `OS Release: ${os__default.default.release()}`,
        `Architecture: ${process.arch}`,
        `CWD: ${process.cwd()}`,
        ""
      ].join("\n");
      await writeToLog(systemInfo);
    },
    async logConfig(config, source) {
      if (!debug) return;
      const sanitizedConfig = {
        buildDir: path3__default.default.basename(config.buildDir),
        minOccurrences: config.minOccurrences,
        minClasses: config.minClasses,
        minBytesSaved: config.minBytesSaved,
        hashPrefix: config.hashPrefix,
        hashLength: config.hashLength,
        cssLayer: config.cssLayer,
        dataAttributes: config.dataAttributes,
        manifest: config.manifest,
        backup: config.backup,
        verbose: config.verbose,
        forceAll: config.forceAll,
        excludeDynamicPatterns: config.excludeDynamicPatterns,
        skipPatternsWithExcludedClasses: config.skipPatternsWithExcludedClasses,
        ssr: config.ssr,
        debug: config.debug,
        sendErrorReports: config.sendErrorReports,
        // Exclude errorReportUrl, exclude patterns, and include patterns for privacy
        excludePrefixCount: config.exclude.prefixes?.length || 0,
        excludeSuffixCount: config.exclude.suffixes?.length || 0,
        excludeClassCount: config.exclude.classes?.length || 0,
        excludePatternCount: config.exclude.patterns?.length || 0,
        excludeFileCount: config.exclude.files?.length || 0,
        dynamicPrefixCount: config.dynamicPrefixes?.length || 0,
        includePatternCount: config.include?.length || 0
      };
      const configInfo = [
        "=== CONFIG ===",
        `Source: ${source}`,
        "Final Config:",
        JSON.stringify(sanitizedConfig, null, 2),
        ""
      ].join("\n");
      await writeToLog(configInfo);
    },
    async logStep(step, details) {
      if (!debug) return;
      let entry = `${timestamp()} ${step}`;
      if (details) {
        const detailLines = Object.entries(details).map(([key, value]) => `  ${key}: ${JSON.stringify(value)}`).join("\n");
        entry += "\n" + detailLines;
      }
      await writeToLog(entry);
    },
    async logTiming(operation, durationMs) {
      if (!debug) return;
      await writeToLog(`${timestamp()} ${operation} completed in ${durationMs}ms`);
    },
    async logError(error, context) {
      if (!debug) return;
      const errorInfo = [
        "",
        "=== ERROR ===",
        `${timestamp()} ${context || "Error occurred"}`,
        `Message: ${error.message}`,
        `Stack trace:`,
        error.stack || "No stack trace available",
        ""
      ].join("\n");
      await writeToLog(errorInfo);
    },
    async close() {
      if (!debug || !logPath) return;
      await writeToLog(`
=== LOG COMPLETE ===
Ended: ${(/* @__PURE__ */ new Date()).toISOString()}
`);
    },
    getLogPath() {
      return logPath;
    }
  };
}
var TIMEOUT_MS = 5e3;
function sanitizeConfig(config) {
  return {
    minOccurrences: config.minOccurrences,
    minClasses: config.minClasses,
    minBytesSaved: config.minBytesSaved,
    hashPrefix: config.hashPrefix,
    hashLength: config.hashLength,
    cssLayer: config.cssLayer,
    dataAttributes: config.dataAttributes,
    manifest: config.manifest,
    backup: config.backup,
    verbose: config.verbose,
    forceAll: config.forceAll,
    excludeDynamicPatterns: config.excludeDynamicPatterns,
    skipPatternsWithExcludedClasses: config.skipPatternsWithExcludedClasses,
    ssr: config.ssr,
    debug: config.debug
    // Excluded for privacy:
    // - buildDir (full path reveals project structure)
    // - errorReportUrl (meta/recursive)
    // - sendErrorReports (meta)
    // - exclude.* (could reveal project patterns)
    // - include (could reveal project patterns)
    // - dynamicPrefixes (could reveal libraries used)
  };
}
async function sendErrorReport(error, options, config, command) {
  if (!options.enabled || !options.url) {
    return;
  }
  const payload = {
    timestamp: (/* @__PURE__ */ new Date()).toISOString(),
    classpressoVersion: getVersion(),
    nodeVersion: process.version,
    os: process.platform,
    platform: os__default.default.release(),
    arch: process.arch,
    errorMessage: error.message,
    errorStack: error.stack,
    config: sanitizeConfig(config),
    command
  };
  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), TIMEOUT_MS);
    await fetch(options.url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "User-Agent": `classpresso/${getVersion()}`
      },
      body: JSON.stringify(payload),
      signal: controller.signal
    });
    clearTimeout(timeoutId);
  } catch {
  }
}

// src/cli/analyze.ts
async function analyzeCommand(options) {
  const config = await loadConfig(options.dir);
  config.minOccurrences = parseInt(options.minOccurrences, 10);
  config.minClasses = parseInt(options.minClasses, 10);
  config.verbose = options.verbose || false;
  config.ssr = options.ssr || false;
  config.debug = options.debug || config.debug || false;
  config.sendErrorReports = options.sendErrorReports || config.sendErrorReports || false;
  config.errorReportUrl = options.errorReportUrl || config.errorReportUrl;
  const logger = createLogger({
    debug: config.debug,
    verbose: config.verbose,
    buildDir: config.buildDir
  });
  if (!options.json) {
    console.log(chalk2__default.default.cyan("\n\u2615 Classpresso - Analyzing build output...\n"));
  }
  try {
    await logger.logSystemInfo();
    await logger.logConfig(config, "analyze command");
    const scanStartTime = Date.now();
    await logger.logStep("Starting scan", { buildDir: config.buildDir });
    const scanResult = await scanBuildOutput(config);
    await logger.logTiming("Scan", Date.now() - scanStartTime);
    await logger.logStep("Scan complete", {
      filesScanned: scanResult.files.length,
      uniquePatterns: scanResult.occurrences.size,
      errors: scanResult.errors.length
    });
    if (scanResult.errors.length > 0 && config.verbose) {
      console.log(chalk2__default.default.yellow("Warnings:"));
      for (const error of scanResult.errors) {
        console.log(chalk2__default.default.yellow(`  - ${error}`));
      }
      console.log();
    }
    const detectStartTime = Date.now();
    await logger.logStep("Detecting consolidatable patterns");
    const candidates = detectConsolidatablePatterns(
      scanResult.occurrences,
      config,
      config.ssr ? scanResult.mergeablePatterns : void 0
    );
    const summary = getPatternSummary(candidates);
    await logger.logTiming("Pattern detection", Date.now() - detectStartTime);
    await logger.logStep("Pattern detection complete", {
      candidatesFound: candidates.length,
      totalBytesSaved: summary.totalBytesSaved
    });
    if (options.json) {
      console.log(JSON.stringify({
        summary,
        candidates: candidates.map((c) => ({
          original: c.classString,
          hashName: c.hashName,
          frequency: c.frequency,
          bytesSaved: c.bytesSaved,
          classes: c.classes,
          excludedClasses: c.excludedClasses
        })),
        files: scanResult.files.length,
        errors: scanResult.errors
      }, null, 2));
      await logger.close();
      return;
    }
    console.log(chalk2__default.default.bold("\u{1F4CA} Analysis Summary\n"));
    console.log(`  Files scanned:        ${chalk2__default.default.cyan(scanResult.files.length)}`);
    console.log(`  Unique patterns:      ${chalk2__default.default.cyan(summary.totalPatterns)}`);
    console.log(`  Total occurrences:    ${chalk2__default.default.cyan(summary.totalOccurrences)}`);
    console.log(`  Avg. frequency:       ${chalk2__default.default.cyan(summary.avgFrequency.toFixed(1))}`);
    console.log(`  Avg. classes/pattern: ${chalk2__default.default.cyan(summary.avgClassesPerPattern.toFixed(1))}`);
    console.log();
    console.log(chalk2__default.default.bold("\u{1F4B0} Potential Savings\n"));
    console.log(`  Patterns to consolidate: ${chalk2__default.default.green(summary.totalPatterns)}`);
    console.log(`  Estimated savings:       ${chalk2__default.default.green(formatBytes(summary.totalBytesSaved))}`);
    console.log();
    if (candidates.length > 0) {
      console.log(chalk2__default.default.bold("\u{1F51D} Top 10 Consolidation Candidates\n"));
      const top10 = candidates.slice(0, 10);
      for (let i = 0; i < top10.length; i++) {
        const c = top10[i];
        console.log(
          `  ${chalk2__default.default.gray(`${i + 1}.`)} ${chalk2__default.default.yellow(c.hashName)} ${chalk2__default.default.gray("(")}${chalk2__default.default.cyan(c.frequency)}x${chalk2__default.default.gray(")")} saves ${chalk2__default.default.green(formatBytes(c.bytesSaved))}`
        );
        console.log(`     ${chalk2__default.default.gray(truncate(c.classString, 60))}`);
        if (c.excludedClasses.length > 0) {
          console.log(`     ${chalk2__default.default.gray("excluded:")} ${chalk2__default.default.yellow(c.excludedClasses.join(" "))}`);
        }
      }
      console.log();
    }
    if (candidates.length === 0) {
      console.log(chalk2__default.default.yellow("No patterns found that meet the consolidation criteria."));
      console.log(chalk2__default.default.gray("Try lowering --min-occurrences or --min-classes"));
    } else {
      console.log(chalk2__default.default.gray(`Run ${chalk2__default.default.cyan("classpresso optimize")} to apply these optimizations.`));
    }
    if (config.debug && logger.getLogPath()) {
      console.log(chalk2__default.default.gray(`Debug log: ${logger.getLogPath()}`));
    }
    console.log();
    await logger.close();
  } catch (err) {
    const error = err instanceof Error ? err : new Error(String(err));
    await logger.logError(error, "analyze command");
    await logger.close();
    if (config.sendErrorReports) {
      await sendErrorReport(error, {
        enabled: true,
        url: config.errorReportUrl
      }, config, "analyze");
    }
    console.error(chalk2__default.default.red(`Error: ${error.message}`));
    process.exit(1);
  }
}
function truncate(str, maxLength) {
  if (str.length <= maxLength) return str;
  return str.substring(0, maxLength - 3) + "...";
}

// src/core/consolidator.ts
init_files();
function createClassMappings(candidates) {
  return candidates.map((candidate) => ({
    original: candidate.classString,
    consolidated: candidate.hashName,
    classes: candidate.classes,
    excludedClasses: candidate.excludedClasses,
    cssDeclarations: "",
    // Will be filled by CSS generator
    frequency: candidate.frequency,
    bytesSaved: candidate.bytesSaved
  }));
}
async function saveMappingManifest(mappings, metrics, config) {
  const manifest = {
    version: "1.0.0",
    tool: "classpresso",
    buildDir: config.buildDir,
    created: (/* @__PURE__ */ new Date()).toISOString(),
    config: {
      minOccurrences: config.minOccurrences,
      minClasses: config.minClasses,
      hashPrefix: config.hashPrefix,
      hashLength: config.hashLength
    },
    mappings,
    metrics
  };
  const manifestPath = path3__default.default.join(config.buildDir, "classpresso-manifest.json");
  await writeFileContent(manifestPath, JSON.stringify(manifest, null, 2));
  return manifestPath;
}
async function loadMappingManifest(buildDir) {
  const manifestPath = path3__default.default.join(buildDir, "classpresso-manifest.json");
  try {
    const { readFileContent: readFileContent2 } = await Promise.resolve().then(() => (init_files(), files_exports));
    const content = await readFileContent2(manifestPath);
    return JSON.parse(content);
  } catch {
    return null;
  }
}

// src/core/css-generator.ts
init_files();
var UTILITY_PROPERTY_MAP = {
  // Display
  "flex": "display: flex",
  "grid": "display: grid",
  "block": "display: block",
  "inline": "display: inline",
  "inline-block": "display: inline-block",
  "hidden": "display: none",
  // Flex direction
  "flex-col": "flex-direction: column",
  "flex-row": "flex-direction: row",
  "flex-col-reverse": "flex-direction: column-reverse",
  "flex-row-reverse": "flex-direction: row-reverse",
  // Flex wrap
  "flex-wrap": "flex-wrap: wrap",
  "flex-nowrap": "flex-wrap: nowrap",
  // Justify content
  "justify-start": "justify-content: flex-start",
  "justify-end": "justify-content: flex-end",
  "justify-center": "justify-content: center",
  "justify-between": "justify-content: space-between",
  "justify-around": "justify-content: space-around",
  "justify-evenly": "justify-content: space-evenly",
  // Align items
  "items-start": "align-items: flex-start",
  "items-end": "align-items: flex-end",
  "items-center": "align-items: center",
  "items-baseline": "align-items: baseline",
  "items-stretch": "align-items: stretch",
  // Text align
  "text-left": "text-align: left",
  "text-center": "text-align: center",
  "text-right": "text-align: right",
  // Font weight
  "font-bold": "font-weight: 700",
  "font-semibold": "font-weight: 600",
  "font-medium": "font-weight: 500",
  "font-normal": "font-weight: 400",
  "font-light": "font-weight: 300",
  // Position
  "relative": "position: relative",
  "absolute": "position: absolute",
  "fixed": "position: fixed",
  "sticky": "position: sticky",
  // Overflow
  "overflow-hidden": "overflow: hidden",
  "overflow-auto": "overflow: auto",
  "overflow-scroll": "overflow: scroll",
  "overflow-visible": "overflow: visible",
  // Cursor
  "cursor-pointer": "cursor: pointer",
  "cursor-default": "cursor: default",
  "cursor-not-allowed": "cursor: not-allowed",
  // Misc
  "truncate": ["overflow: hidden", "text-overflow: ellipsis", "white-space: nowrap"],
  "uppercase": "text-transform: uppercase",
  "lowercase": "text-transform: lowercase",
  "capitalize": "text-transform: capitalize"
};
function parseUtilityClass(className) {
  if (UTILITY_PROPERTY_MAP[className]) {
    const value = UTILITY_PROPERTY_MAP[className];
    return Array.isArray(value) ? value : [value];
  }
  const arbitraryMatch = className.match(/^([a-z]+(?:-[a-z]+)?)-\[(.+)\]$/);
  if (arbitraryMatch) {
    const [, prefix, value] = arbitraryMatch;
    return parseArbitraryValue(prefix, value);
  }
  const spacingMatch = className.match(/^(p|px|py|pt|pb|pl|pr|m|mx|my|mt|mb|ml|mr|gap|gap-x|gap-y)-(\d+(?:\.\d+)?|\[.+\])$/);
  if (spacingMatch) {
    const [, prefix, value] = spacingMatch;
    return parseSpacingUtility(prefix, value);
  }
  const sizeMatch = className.match(/^(w|h|min-w|min-h|max-w|max-h)-(.+)$/);
  if (sizeMatch) {
    const [, prefix, value] = sizeMatch;
    return parseSizeUtility(prefix, value);
  }
  const roundedMatch = className.match(/^rounded(-[a-z]+)?(-[a-z0-9]+)?$/);
  if (roundedMatch) {
    return parseRoundedUtility(className);
  }
  const textSizeMatch = className.match(/^text-(xs|sm|base|lg|xl|2xl|3xl|4xl|5xl|6xl|7xl|8xl|9xl)$/);
  if (textSizeMatch) {
    return parseTextSizeUtility(textSizeMatch[1]);
  }
  const opacityMatch = className.match(/^opacity-(\d+)$/);
  if (opacityMatch) {
    return [`opacity: ${parseInt(opacityMatch[1]) / 100}`];
  }
  const zMatch = className.match(/^z-(\d+|auto)$/);
  if (zMatch) {
    return [`z-index: ${zMatch[1]}`];
  }
  return [];
}
function parseArbitraryValue(prefix, value) {
  const propertyMap = {
    "text": "color",
    "bg": "background-color",
    "border": "border-color",
    "border-t": "border-top-color",
    "border-b": "border-bottom-color",
    "border-l": "border-left-color",
    "border-r": "border-right-color",
    "fill": "fill",
    "stroke": "stroke",
    "p": "padding",
    "px": "padding-inline",
    "py": "padding-block",
    "pt": "padding-top",
    "pb": "padding-bottom",
    "pl": "padding-left",
    "pr": "padding-right",
    "m": "margin",
    "mx": "margin-inline",
    "my": "margin-block",
    "mt": "margin-top",
    "mb": "margin-bottom",
    "ml": "margin-left",
    "mr": "margin-right",
    "w": "width",
    "h": "height",
    "min-w": "min-width",
    "min-h": "min-height",
    "max-w": "max-width",
    "max-h": "max-height",
    "top": "top",
    "bottom": "bottom",
    "left": "left",
    "right": "right",
    "gap": "gap",
    "gap-x": "column-gap",
    "gap-y": "row-gap"
  };
  const property = propertyMap[prefix];
  if (property) {
    return [`${property}: ${value}`];
  }
  return [];
}
function parseSpacingUtility(prefix, value) {
  const spacingScale = {
    "0": "0px",
    "0.5": "0.125rem",
    "1": "0.25rem",
    "1.5": "0.375rem",
    "2": "0.5rem",
    "2.5": "0.625rem",
    "3": "0.75rem",
    "3.5": "0.875rem",
    "4": "1rem",
    "5": "1.25rem",
    "6": "1.5rem",
    "7": "1.75rem",
    "8": "2rem",
    "9": "2.25rem",
    "10": "2.5rem",
    "11": "2.75rem",
    "12": "3rem",
    "14": "3.5rem",
    "16": "4rem",
    "20": "5rem",
    "24": "6rem",
    "28": "7rem",
    "32": "8rem",
    "36": "9rem",
    "40": "10rem",
    "44": "11rem",
    "48": "12rem",
    "52": "13rem",
    "56": "14rem",
    "60": "15rem",
    "64": "16rem",
    "72": "18rem",
    "80": "20rem",
    "96": "24rem"
  };
  let cssValue = spacingScale[value];
  if (!cssValue && value.startsWith("[") && value.endsWith("]")) {
    cssValue = value.slice(1, -1);
  }
  if (!cssValue) return [];
  const propertyMap = {
    "p": "padding",
    "px": ["padding-left", "padding-right"],
    "py": ["padding-top", "padding-bottom"],
    "pt": "padding-top",
    "pb": "padding-bottom",
    "pl": "padding-left",
    "pr": "padding-right",
    "m": "margin",
    "mx": ["margin-left", "margin-right"],
    "my": ["margin-top", "margin-bottom"],
    "mt": "margin-top",
    "mb": "margin-bottom",
    "ml": "margin-left",
    "mr": "margin-right",
    "gap": "gap",
    "gap-x": "column-gap",
    "gap-y": "row-gap"
  };
  const property = propertyMap[prefix];
  if (Array.isArray(property)) {
    return property.map((p) => `${p}: ${cssValue}`);
  }
  if (property) {
    return [`${property}: ${cssValue}`];
  }
  return [];
}
function parseSizeUtility(prefix, value) {
  const sizeMap = {
    "full": "100%",
    "screen": "100vh",
    "svh": "100svh",
    "lvh": "100lvh",
    "dvh": "100dvh",
    "min": "min-content",
    "max": "max-content",
    "fit": "fit-content",
    "auto": "auto"
  };
  let cssValue = sizeMap[value];
  if (!cssValue && value.startsWith("[") && value.endsWith("]")) {
    cssValue = value.slice(1, -1);
  }
  if (!cssValue) {
    const spacingMatch = value.match(/^(\d+(?:\.\d+)?)$/);
    if (spacingMatch) {
      cssValue = `${parseFloat(spacingMatch[1]) * 0.25}rem`;
    }
  }
  if (!cssValue) return [];
  const propertyMap = {
    "w": "width",
    "h": "height",
    "min-w": "min-width",
    "min-h": "min-height",
    "max-w": "max-width",
    "max-h": "max-height"
  };
  const property = propertyMap[prefix];
  if (property) {
    return [`${property}: ${cssValue}`];
  }
  return [];
}
function parseRoundedUtility(className) {
  const roundedMap = {
    "rounded": "border-radius: 0.25rem",
    "rounded-none": "border-radius: 0px",
    "rounded-sm": "border-radius: 0.125rem",
    "rounded-md": "border-radius: 0.375rem",
    "rounded-lg": "border-radius: 0.5rem",
    "rounded-xl": "border-radius: 0.75rem",
    "rounded-2xl": "border-radius: 1rem",
    "rounded-3xl": "border-radius: 1.5rem",
    "rounded-full": "border-radius: 9999px"
  };
  if (roundedMap[className]) {
    return [roundedMap[className]];
  }
  return [];
}
function parseTextSizeUtility(size) {
  const sizeMap = {
    "xs": ["0.75rem", "1rem"],
    "sm": ["0.875rem", "1.25rem"],
    "base": ["1rem", "1.5rem"],
    "lg": ["1.125rem", "1.75rem"],
    "xl": ["1.25rem", "1.75rem"],
    "2xl": ["1.5rem", "2rem"],
    "3xl": ["1.875rem", "2.25rem"],
    "4xl": ["2.25rem", "2.5rem"],
    "5xl": ["3rem", "1"],
    "6xl": ["3.75rem", "1"],
    "7xl": ["4.5rem", "1"],
    "8xl": ["6rem", "1"],
    "9xl": ["8rem", "1"]
  };
  const [fontSize, lineHeight] = sizeMap[size] || ["1rem", "1.5rem"];
  return [`font-size: ${fontSize}`, `line-height: ${lineHeight}`];
}
async function extractUtilityCSS(buildDir) {
  const utilityMap = /* @__PURE__ */ new Map();
  const cssPatterns = [
    // Next.js
    "static/**/*.css",
    "static/css/**/*.css",
    // Angular
    "browser/**/*.css",
    // Astro
    "_astro/**/*.css",
    // Nuxt
    "public/_nuxt/**/*.css",
    // Vite/Vue/React/Qwik generic
    "assets/**/*.css",
    // Gatsby/Hugo/Eleventy/Docusaurus/Parcel
    "**/*.css"
  ];
  const cssFiles = await findFiles(buildDir, cssPatterns);
  for (const cssFile of cssFiles) {
    if (!isCSSFile(cssFile)) continue;
    try {
      const content = await readFileContent(cssFile);
      const root = postcss__default.default.parse(content);
      root.walkRules((rule) => {
        const selectorMatch = rule.selector.match(/^\.([^\s:,\[]+)/);
        if (!selectorMatch) return;
        const className = selectorMatch[1];
        const declarations = [];
        rule.walkDecls((decl) => {
          declarations.push(`${decl.prop}: ${decl.value}`);
        });
        if (declarations.length > 0) {
          utilityMap.set(className, declarations);
        }
      });
    } catch {
    }
  }
  return utilityMap;
}
async function generateConsolidatedCSS(mappings, buildDir, cssLayer) {
  const utilityMap = await extractUtilityCSS(buildDir);
  const cssRules = [];
  for (const mapping of mappings) {
    const declarations = [];
    for (const className of mapping.classes) {
      const extracted = utilityMap.get(className);
      if (extracted) {
        declarations.push(...extracted);
        continue;
      }
      const parsed = parseUtilityClass(className);
      if (parsed.length > 0) {
        declarations.push(...parsed);
      }
    }
    const uniqueDeclarations = [...new Set(declarations)];
    if (uniqueDeclarations.length > 0) {
      mapping.cssDeclarations = uniqueDeclarations.join("; ");
      cssRules.push(`.${mapping.consolidated} {
  ${uniqueDeclarations.join(";\n  ")};
}`);
    }
  }
  const cssContent = cssRules.join("\n\n");
  if (cssLayer) {
    return `/* Classpresso Consolidated Classes */
@layer ${cssLayer} {
${cssRules.map((rule) => "  " + rule.replace(/\n/g, "\n  ")).join("\n\n")}
}`;
  }
  return `/* Classpresso Consolidated Classes */
${cssContent}`;
}
async function injectConsolidatedCSS(buildDir, consolidatedCSS) {
  const cssPatterns = [
    // Next.js
    "static/**/*.css",
    "static/css/**/*.css",
    "standalone/.next/static/**/*.css",
    "standalone/.next/static/css/**/*.css",
    // Angular
    "browser/**/*.css",
    // Astro
    "_astro/**/*.css",
    "client/_astro/**/*.css",
    // Nuxt
    "public/_nuxt/**/*.css",
    // SvelteKit
    "_app/**/*.css",
    "client/**/*.css",
    // Vite/Vue/React/Qwik generic
    "assets/**/*.css",
    // Gatsby/Hugo/Eleventy/Docusaurus/Parcel
    "**/*.css"
  ];
  const cssFiles = await findFiles(buildDir, cssPatterns);
  if (cssFiles.length === 0) {
    throw new Error("No CSS files found in build output");
  }
  let targetFile = cssFiles[0];
  let maxSize = 0;
  for (const cssFile of cssFiles) {
    if (!isCSSFile(cssFile)) continue;
    try {
      const content = await readFileContent(cssFile);
      if (content.includes("Classpresso Consolidated")) {
        return `none (already injected in ${cssFile})`;
      }
      const size = Buffer.byteLength(content, "utf-8");
      if (size > maxSize) {
        maxSize = size;
        targetFile = cssFile;
      }
    } catch {
    }
  }
  const existingContent = await readFileContent(targetFile);
  const newContent = `${existingContent}

${consolidatedCSS}`;
  await writeFileContent(targetFile, newContent);
  return targetFile;
}

// src/core/transformer.ts
init_files();
function isSupersetOfDynamicBase(classes, dynamicBasePatterns) {
  const classSet = new Set(classes);
  for (const [, pattern] of dynamicBasePatterns) {
    const allBasePresent = pattern.baseClasses.every((cls) => classSet.has(cls));
    if (allBasePresent && classes.length > pattern.baseClasses.length) {
      return true;
    }
  }
  return false;
}
function matchesDynamicBase(classes, dynamicBasePatterns) {
  const classSet = new Set(classes);
  for (const [, pattern] of dynamicBasePatterns) {
    if (classes.length !== pattern.baseClasses.length) continue;
    const allMatch = pattern.baseClasses.every((cls) => classSet.has(cls));
    if (allMatch) return true;
  }
  return false;
}
function buildMatchPatterns(classString) {
  const escaped = escapeRegex(classString);
  return [
    // className="..." (JSX/HTML) - supports data attributes
    { regex: new RegExp(`(className\\s*=\\s*")${escaped}(")`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(className\\s*=\\s*')${escaped}(')`, "g"), supportsDataAttr: true },
    // className:"..." (React createElement) - no data attr support
    { regex: new RegExp(`(className\\s*:\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    { regex: new RegExp(`(className\\s*:\\s*')${escaped}(')`, "g"), supportsDataAttr: false },
    // Template literals: className:`...` (static content only)
    { regex: new RegExp(`(className\\s*:\\s*\`)${escaped}(\`)`, "g"), supportsDataAttr: false },
    // Template literals with dynamic suffix: className:`... ${...}`
    // Matches static prefix before ${...} expression
    { regex: new RegExp(`(className\\s*:\\s*\`)${escaped}( \\$\\{)`, "g"), supportsDataAttr: false },
    // "className","..." (minified) - no data attr support
    { regex: new RegExp(`("className"\\s*,\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    // class="..." (HTML) - supports data attributes
    { regex: new RegExp(`(\\bclass\\s*=\\s*")${escaped}(")`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(\\bclass\\s*=\\s*')${escaped}(')`, "g"), supportsDataAttr: true },
    // class:"..." (Vue/Quasar createElement) - no data attr support
    { regex: new RegExp(`(\\bclass\\s*:\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    { regex: new RegExp(`(\\bclass\\s*:\\s*')${escaped}(')`, "g"), supportsDataAttr: false },
    // HTML entity encoded: class=&quot;...&quot; - supports data attributes
    { regex: new RegExp(`(class=&quot;)${escaped}(&quot;)`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(class=&#34;)${escaped}(&#34;)`, "g"), supportsDataAttr: true },
    // RSC payload (escaped JSON in HTML): \"className\":\"...\"
    // This is used in Next.js self.__next_f.push() payloads
    { regex: new RegExp(`(\\\\"className\\\\"\\s*:\\s*\\\\")${escaped}(\\\\")`, "g"), supportsDataAttr: false }
  ];
}
function replacePattern(content, originalClasses, consolidated, excludedClasses, originalClassList, addDataAttributes) {
  let replacements = 0;
  let result = content;
  const replacement = excludedClasses.length > 0 ? `${consolidated} ${excludedClasses.join(" ")}` : consolidated;
  const dataAttr = addDataAttributes ? ` data-cp-original="${originalClassList.join(" ")}"` : "";
  const patterns = buildMatchPatterns(originalClasses);
  for (const { regex: pattern, supportsDataAttr } of patterns) {
    const beforeLength = result.length;
    const suffix = addDataAttributes && supportsDataAttr ? dataAttr : "";
    result = result.replace(pattern, `$1${replacement}$2${suffix}`);
    if (result.length !== beforeLength) {
      const matches = content.match(pattern);
      if (matches) replacements += matches.length;
    }
  }
  return { content: result, replacements };
}
function findClassVariations(content, mapping, config) {
  const variations = [];
  const targetClasses = new Set(mapping.classes);
  const classAttrPatterns = [
    /(?:className|class)\s*[=:]\s*["']([^"']+)["']/g,
    /class=&quot;([^&]+)&quot;/g,
    /class=&#34;([^&]+)&#34;/g,
    // Template literals - static only: className:`...`
    /(?:className|class)\s*:\s*`([^`$]+)`/g,
    // Template literals with dynamic suffix: className:`... ${...}`
    // Captures the static part before the ${
    /(?:className|class)\s*:\s*`([^`$]+) \$\{/g,
    // RSC payload (escaped JSON in HTML): \"className\":\"...\"
    // Used in Next.js self.__next_f.push() payloads
    /\\"className\\"\s*:\s*\\"([^"\\]+)\\"/g
  ];
  for (const classAttrRegex of classAttrPatterns) {
    classAttrRegex.lastIndex = 0;
    let match;
    while ((match = classAttrRegex.exec(content)) !== null) {
      const classString = match[1];
      const { classes } = normalizeClassString(classString, config.exclude);
      if (classes.length >= targetClasses.size) {
        const classSet = new Set(classes);
        let allPresent = true;
        for (const targetClass of targetClasses) {
          if (!classSet.has(targetClass)) {
            allPresent = false;
            break;
          }
        }
        if (allPresent && classes.length === targetClasses.size) {
          variations.push(classString);
        }
      }
    }
  }
  return [...new Set(variations)];
}
async function transformFile(filePath, mappings, config, dryRun, dynamicBasePatterns, mergeablePatterns) {
  try {
    let content = await readFileContent(filePath);
    let totalReplacements = 0;
    let skippedForHydration = 0;
    const originalContent = content;
    const isJS = isJSFile(filePath);
    for (const mapping of mappings) {
      if (dynamicBasePatterns.size > 0) {
        const isSuperset = isSupersetOfDynamicBase(mapping.classes, dynamicBasePatterns);
        const isExact = matchesDynamicBase(mapping.classes, dynamicBasePatterns);
        if (isSuperset || isExact) {
          skippedForHydration++;
          continue;
        }
      }
      if (isJS && mergeablePatterns.has(mapping.original)) {
        skippedForHydration++;
        continue;
      }
      const variations = findClassVariations(content, mapping, config);
      for (const variation of variations) {
        const { content: newContent, replacements } = replacePattern(
          content,
          variation,
          mapping.consolidated,
          mapping.excludedClasses,
          mapping.classes,
          config.dataAttributes
        );
        content = newContent;
        totalReplacements += replacements;
      }
    }
    const modified = content !== originalContent;
    if (modified && !dryRun) {
      if (config.backup) {
        await backupFile(filePath);
      }
      await writeFileContent(filePath, content);
    }
    return { modified, replacements: totalReplacements, skippedForHydration };
  } catch (err) {
    return {
      modified: false,
      replacements: 0,
      skippedForHydration: 0,
      error: `Error transforming ${filePath}: ${err}`
    };
  }
}
async function transformBuildOutput(mappings, config, dryRun = false, dynamicBasePatterns = /* @__PURE__ */ new Map(), mergeablePatterns = /* @__PURE__ */ new Set()) {
  if (config.verbose) {
    if (dynamicBasePatterns.size > 0) {
      console.log(`  \u2139 Found ${dynamicBasePatterns.size} dynamic base patterns (template literals)`);
    }
    if (mergeablePatterns.size > 0) {
      console.log(`  \u2139 Found ${mergeablePatterns.size} mergeable patterns (className props)`);
    }
  }
  const patterns = config.include.length > 0 ? config.include : DEFAULT_PATTERNS;
  const allFiles = await findFiles(config.buildDir, patterns);
  const files = filterExcludedFiles(allFiles, config.exclude.files || []);
  if (config.verbose && allFiles.length !== files.length) {
    console.log(`  \u2139 Excluded ${allFiles.length - files.length} files from transformation`);
  }
  let filesModified = 0;
  let totalBytesChanged = 0;
  let totalSkippedForHydration = 0;
  const errors = [];
  for (const filePath of files) {
    if (!isJSFile(filePath) && !isHTMLFile(filePath) && !isRSCFile(filePath)) {
      continue;
    }
    const result = await transformFile(filePath, mappings, config, dryRun, dynamicBasePatterns, mergeablePatterns);
    if (result.error) {
      errors.push(result.error);
    }
    if (result.modified) {
      filesModified++;
      totalBytesChanged += result.replacements * 10;
    }
    totalSkippedForHydration += result.skippedForHydration;
  }
  if (config.verbose && totalSkippedForHydration > 0) {
    console.log(`Skipped ${totalSkippedForHydration} pattern transformations to prevent hydration mismatches`);
  }
  return {
    filesModified,
    bytesChanged: totalBytesChanged,
    errors
  };
}

// src/core/css-purger.ts
init_files();
var CSS_PATTERNS = [
  // Next.js
  "static/**/*.css",
  "static/css/**/*.css",
  "standalone/.next/static/**/*.css",
  // Angular
  "browser/**/*.css",
  // Astro
  "_astro/**/*.css",
  "client/_astro/**/*.css",
  // Nuxt
  "public/_nuxt/**/*.css",
  // SvelteKit
  "_app/**/*.css",
  "client/**/*.css",
  // Vite/Vue/React/Qwik generic
  "assets/**/*.css",
  // Generic fallback
  "**/*.css"
];
var SCAN_PATTERNS = [
  "**/*.html",
  "**/*.js",
  "**/*.mjs",
  "**/*.jsx"
];
async function extractUsedClasses(buildDir) {
  const usedClasses = /* @__PURE__ */ new Set();
  const files = await findFiles(buildDir, SCAN_PATTERNS);
  for (const file of files) {
    try {
      const content = await readFileContent(file);
      for (const pattern of ALL_CLASS_PATTERNS) {
        const regex = new RegExp(pattern.source, pattern.flags);
        let match;
        while ((match = regex.exec(content)) !== null) {
          const classString = match[1] || match[2] || match[3];
          if (classString) {
            const classes = classString.split(/\s+/).filter((c) => c.length > 0);
            classes.forEach((cls) => usedClasses.add(cls));
          }
        }
      }
    } catch {
    }
  }
  return usedClasses;
}
function extractClassFromSelector(selector) {
  const match = selector.match(/^\.([^\s:,\[]+)/);
  if (match) {
    return match[1].replace(/\\/g, "");
  }
  return null;
}
function matchesSafelist(className, safelist) {
  for (const pattern of safelist) {
    if (typeof pattern === "string") {
      if (className === pattern) return true;
    } else if (pattern instanceof RegExp) {
      if (pattern.test(className)) return true;
    }
  }
  return false;
}
function isResponsiveOrStateVariant(className) {
  const variantPrefixes = [
    "sm:",
    "md:",
    "lg:",
    "xl:",
    "2xl:",
    // Responsive
    "hover:",
    "focus:",
    "active:",
    "disabled:",
    "group-hover:",
    // State
    "dark:",
    "light:",
    // Theme
    "first:",
    "last:",
    "odd:",
    "even:",
    // Child
    "before:",
    "after:"
    // Pseudo-elements
  ];
  return variantPrefixes.some((prefix) => className.startsWith(prefix));
}
async function purgeUnusedCSS(buildDir, mappings, config) {
  const result = {
    filesProcessed: 0,
    rulesRemoved: 0,
    purgedClasses: [],
    bytesSaved: 0,
    keptClasses: [],
    errors: []
  };
  try {
    const consolidatedClasses = /* @__PURE__ */ new Set();
    for (const mapping of mappings) {
      mapping.classes.forEach((cls) => consolidatedClasses.add(cls));
      consolidatedClasses.add(mapping.consolidated);
    }
    const usedClasses = await extractUsedClasses(buildDir);
    const cssFiles = await findFiles(buildDir, CSS_PATTERNS);
    for (const cssFile of cssFiles) {
      if (!isCSSFile(cssFile)) continue;
      try {
        const originalContent = await readFileContent(cssFile);
        const originalSize = Buffer.byteLength(originalContent, "utf-8");
        const root = postcss__default.default.parse(originalContent);
        const rulesToRemove = [];
        root.walkRules((rule) => {
          const className = extractClassFromSelector(rule.selector);
          if (!className) return;
          if (usedClasses.has(className)) {
            result.keptClasses.push(className);
            return;
          }
          if (className.startsWith(config.hashPrefix)) {
            return;
          }
          if (config.purgeSafelist && matchesSafelist(className, config.purgeSafelist)) {
            result.keptClasses.push(className);
            return;
          }
          if (isResponsiveOrStateVariant(className)) {
            result.keptClasses.push(className);
            return;
          }
          rulesToRemove.push(rule);
          result.purgedClasses.push(className);
        });
        for (const rule of rulesToRemove) {
          rule.remove();
          result.rulesRemoved++;
        }
        if (rulesToRemove.length > 0) {
          const newContent = root.toString();
          const newSize = Buffer.byteLength(newContent, "utf-8");
          result.bytesSaved += originalSize - newSize;
          await writeFileContent(cssFile, newContent);
        }
        result.filesProcessed++;
      } catch (error) {
        result.errors.push(`Error processing ${cssFile}: ${error}`);
      }
    }
    result.purgedClasses = [...new Set(result.purgedClasses)];
    result.keptClasses = [...new Set(result.keptClasses)];
  } catch (error) {
    result.errors.push(`Purge error: ${error}`);
  }
  return result;
}

// src/cli/optimize.ts
async function optimizeCommand(options) {
  const config = await loadConfig(options.dir);
  config.minOccurrences = parseInt(options.minOccurrences, 10);
  config.minClasses = parseInt(options.minClasses, 10);
  config.ssr = options.ssr || false;
  config.backup = options.backup || false;
  config.manifest = options.manifest !== false;
  config.verbose = options.verbose || false;
  config.debug = options.debug || config.debug || false;
  config.sendErrorReports = options.sendErrorReports || config.sendErrorReports || false;
  config.errorReportUrl = options.errorReportUrl || config.errorReportUrl;
  config.purgeUnusedCSS = options.purgeUnused || config.purgeUnusedCSS || false;
  const dryRun = options.dryRun || false;
  const logger = createLogger({
    debug: config.debug,
    verbose: config.verbose,
    buildDir: config.buildDir
  });
  console.log(chalk2__default.default.cyan("\n\u2615 Classpresso - Optimizing build output...\n"));
  if (dryRun) {
    console.log(chalk2__default.default.yellow("\u{1F50D} DRY RUN MODE - No files will be modified\n"));
  }
  if (config.ssr) {
    console.log(chalk2__default.default.blue("\u{1F512} SSR MODE - Only transforming patterns found in both HTML and JS\n"));
  }
  try {
    await logger.logSystemInfo();
    await logger.logConfig(config, "optimize command");
    console.log(chalk2__default.default.gray("Scanning build output..."));
    const scanStartTime = Date.now();
    await logger.logStep("Starting scan", { buildDir: config.buildDir, dryRun });
    const scanResult = await scanBuildOutput(config);
    await logger.logTiming("Scan", Date.now() - scanStartTime);
    await logger.logStep("Scan complete", {
      filesScanned: scanResult.files.length,
      uniquePatterns: scanResult.occurrences.size,
      errors: scanResult.errors.length
    });
    console.log(chalk2__default.default.green(`  \u2713 Scanned ${scanResult.files.length} files
`));
    console.log(chalk2__default.default.gray("Detecting patterns..."));
    const detectStartTime = Date.now();
    await logger.logStep("Detecting patterns");
    const candidates = detectConsolidatablePatterns(
      scanResult.occurrences,
      config,
      config.ssr ? scanResult.mergeablePatterns : void 0
    );
    await logger.logTiming("Pattern detection", Date.now() - detectStartTime);
    await logger.logStep("Pattern detection complete", { candidatesFound: candidates.length });
    console.log(chalk2__default.default.green(`  \u2713 Found ${candidates.length} patterns to consolidate
`));
    if (candidates.length === 0) {
      console.log(chalk2__default.default.yellow("No patterns found that meet the consolidation criteria."));
      console.log(chalk2__default.default.gray("Try lowering --min-occurrences or --min-classes\n"));
      await logger.logStep("No patterns to consolidate - exiting");
      await logger.close();
      return;
    }
    console.log(chalk2__default.default.gray("Creating class mappings..."));
    const mappingStartTime = Date.now();
    await logger.logStep("Creating class mappings");
    const mappings = createClassMappings(candidates);
    await logger.logTiming("Mapping creation", Date.now() - mappingStartTime);
    await logger.logStep("Mappings created", { mappingCount: mappings.length });
    console.log(chalk2__default.default.green(`  \u2713 Created ${mappings.length} mappings
`));
    console.log(chalk2__default.default.gray("Generating consolidated CSS..."));
    const cssStartTime = Date.now();
    await logger.logStep("Generating consolidated CSS");
    const consolidatedCSS = await generateConsolidatedCSS(mappings, config.buildDir, config.cssLayer);
    const cssBytes = Buffer.byteLength(consolidatedCSS, "utf-8");
    await logger.logTiming("CSS generation", Date.now() - cssStartTime);
    await logger.logStep("CSS generated", { cssBytes });
    console.log(chalk2__default.default.green(`  \u2713 Generated ${formatBytes(cssBytes)} of CSS
`));
    console.log(chalk2__default.default.gray("Transforming build output..."));
    const transformStartTime = Date.now();
    await logger.logStep("Transforming build output", { dryRun });
    const transformResult = await transformBuildOutput(
      mappings,
      config,
      dryRun,
      scanResult.dynamicBasePatterns,
      scanResult.mergeablePatterns
    );
    await logger.logTiming("Transformation", Date.now() - transformStartTime);
    await logger.logStep("Transformation complete", {
      filesModified: transformResult.filesModified,
      bytesChanged: transformResult.bytesChanged,
      errors: transformResult.errors.length
    });
    console.log(chalk2__default.default.green(`  \u2713 Modified ${transformResult.filesModified} files
`));
    if (config.verbose) {
      if (scanResult.dynamicBasePatterns.size > 0) {
        console.log(chalk2__default.default.gray(`  \u2139 Found ${scanResult.dynamicBasePatterns.size} dynamic class patterns (template literals)
`));
      }
      if (scanResult.mergeablePatterns.size > 0) {
        console.log(chalk2__default.default.gray(`  \u2139 Found ${scanResult.mergeablePatterns.size} mergeable patterns (className props)
`));
      }
    }
    if (!dryRun) {
      console.log(chalk2__default.default.gray("Injecting consolidated CSS..."));
      const injectStartTime = Date.now();
      await logger.logStep("Injecting consolidated CSS");
      const cssFile = await injectConsolidatedCSS(config.buildDir, consolidatedCSS);
      await logger.logTiming("CSS injection", Date.now() - injectStartTime);
      await logger.logStep("CSS injected", { cssFile });
      console.log(chalk2__default.default.green(`  \u2713 Injected into ${cssFile}
`));
    }
    let purgeBytesSaved = 0;
    if (config.purgeUnusedCSS && !dryRun) {
      console.log(chalk2__default.default.gray("Purging unused CSS..."));
      const purgeStartTime = Date.now();
      await logger.logStep("Purging unused CSS");
      const purgeResult = await purgeUnusedCSS(config.buildDir, mappings, config);
      await logger.logTiming("CSS purge", Date.now() - purgeStartTime);
      await logger.logStep("CSS purge complete", {
        rulesRemoved: purgeResult.rulesRemoved,
        bytesSaved: purgeResult.bytesSaved
      });
      purgeBytesSaved = purgeResult.bytesSaved;
      console.log(chalk2__default.default.green(`  \u2713 Purged ${purgeResult.rulesRemoved} unused CSS rules (${formatBytes(purgeResult.bytesSaved)} saved)
`));
      if (purgeResult.errors.length > 0 && config.verbose) {
        for (const error of purgeResult.errors) {
          console.log(chalk2__default.default.yellow(`  \u26A0 ${error}`));
        }
      }
    }
    await logger.logStep("Calculating metrics");
    const cssOverhead = estimateCSSOverhead(candidates);
    const metrics = calculateMetrics(candidates, scanResult.files, transformResult, cssOverhead);
    await logger.logStep("Metrics calculated", {
      bytesSaved: metrics.bytesSaved,
      percentageReduction: metrics.percentageReduction
    });
    if (!dryRun && config.manifest) {
      console.log(chalk2__default.default.gray("Saving manifest..."));
      await logger.logStep("Saving manifest");
      const manifestPath = await saveMappingManifest(mappings, metrics, config);
      await logger.logStep("Manifest saved", { manifestPath });
      console.log(chalk2__default.default.green(`  \u2713 Saved to ${manifestPath}
`));
    }
    printResults(metrics, dryRun);
    if (transformResult.errors.length > 0) {
      console.log(chalk2__default.default.yellow("\nWarnings:"));
      for (const error of transformResult.errors) {
        console.log(chalk2__default.default.yellow(`  - ${error}`));
      }
    }
    if (config.debug && logger.getLogPath()) {
      console.log(chalk2__default.default.gray(`Debug log: ${logger.getLogPath()}`));
    }
    await logger.close();
  } catch (err) {
    const error = err instanceof Error ? err : new Error(String(err));
    await logger.logError(error, "optimize command");
    await logger.close();
    if (config.sendErrorReports) {
      await sendErrorReport(error, {
        enabled: true,
        url: config.errorReportUrl
      }, config, "optimize");
    }
    console.error(chalk2__default.default.red(`
Error: ${error.message}`));
    process.exit(1);
  }
}
function printResults(metrics, dryRun) {
  const title = dryRun ? "CLASSPRESSO ANALYSIS" : "CLASSPRESSO RESULTS";
  const borderChar = "\u2550";
  const width = 60;
  console.log(chalk2__default.default.cyan(`
\u2554${borderChar.repeat(width)}\u2557`));
  console.log(chalk2__default.default.cyan(`\u2551${" ".repeat((width - title.length) / 2)}${title}${" ".repeat((width - title.length) / 2)}\u2551`));
  console.log(chalk2__default.default.cyan(`\u2560${borderChar.repeat(width)}\u2563`));
  const lines = [
    ["Files Scanned:", metrics.totalFilesScanned.toString()],
    ["Files Modified:", metrics.totalFilesModified.toString()],
    ["", ""],
    ["CLASS ANALYSIS", ""],
    ["Total class strings:", metrics.totalClassStringsFound.toString()],
    ["Unique patterns:", metrics.uniqueClassPatterns.toString()],
    ["Consolidated:", metrics.consolidatedPatterns.toString()],
    ["", ""],
    ["SIZE IMPACT", ""],
    ["Bytes saved (JS/HTML):", formatBytes(metrics.bytesSaved)],
    ["CSS overhead:", formatBytes(metrics.consolidatedCSSBytes)],
    ["Net reduction:", `${formatBytes(metrics.bytesSaved - metrics.consolidatedCSSBytes)} (${formatPercentage(metrics.percentageReduction)})`],
    ["", ""],
    ["BROWSER IMPACT", ""],
    ["Parse time saved:", formatTime(metrics.estimatedParseTimeSavedMs)],
    ["Render time saved:", formatTime(metrics.estimatedRenderTimeSavedMs)]
  ];
  for (const [label, value] of lines) {
    if (label === "") {
      console.log(chalk2__default.default.cyan(`\u2560${borderChar.repeat(width)}\u2563`));
    } else if (value === "") {
      const padding = width - label.length - 2;
      console.log(chalk2__default.default.cyan(`\u2551 ${chalk2__default.default.bold(label)}${" ".repeat(padding)} \u2551`));
    } else {
      const padding = width - label.length - value.length - 3;
      console.log(chalk2__default.default.cyan(`\u2551 ${label}${" ".repeat(Math.max(1, padding))}${chalk2__default.default.green(value)} \u2551`));
    }
  }
  console.log(chalk2__default.default.cyan(`\u2560${borderChar.repeat(width)}\u2563`));
  console.log(chalk2__default.default.cyan(`\u2551 TOP CONSOLIDATIONS${" ".repeat(width - 20)} \u2551`));
  for (let i = 0; i < Math.min(5, metrics.topConsolidations.length); i++) {
    const c = metrics.topConsolidations[i];
    const line = ` ${i + 1}. ${c.consolidated} (${c.frequency}x) saves ${formatBytes(c.bytesSaved)}`;
    const padding = width - line.length - 1;
    console.log(chalk2__default.default.cyan(`\u2551${line}${" ".repeat(Math.max(1, padding))}\u2551`));
  }
  console.log(chalk2__default.default.cyan(`\u255A${borderChar.repeat(width)}\u255D
`));
}
async function reportCommand(options) {
  const manifest = await loadMappingManifest(options.dir);
  if (!manifest) {
    console.error(chalk2__default.default.red("No classpresso manifest found in build directory."));
    console.log(chalk2__default.default.gray('Run "classpresso optimize" first to generate a manifest.'));
    process.exit(1);
  }
  switch (options.format) {
    case "json":
      console.log(JSON.stringify(manifest, null, 2));
      break;
    case "html":
      console.log(generateHTMLReport(manifest));
      break;
    case "text":
    default:
      printTextReport(manifest);
      break;
  }
}
function printTextReport(manifest) {
  const { metrics } = manifest;
  console.log(chalk2__default.default.cyan("\n\u2615 CLASSPRESSO REPORT\n"));
  console.log(chalk2__default.default.gray(`Generated: ${manifest.created}`));
  console.log(chalk2__default.default.gray(`Build: ${manifest.buildDir}
`));
  console.log(chalk2__default.default.bold("\u{1F4CA} Summary\n"));
  console.log(`  Files scanned:     ${chalk2__default.default.cyan(metrics.totalFilesScanned)}`);
  console.log(`  Files modified:    ${chalk2__default.default.cyan(metrics.totalFilesModified)}`);
  console.log(`  Patterns found:    ${chalk2__default.default.cyan(metrics.uniqueClassPatterns)}`);
  console.log(`  Patterns applied:  ${chalk2__default.default.cyan(metrics.consolidatedPatterns)}`);
  console.log();
  console.log(chalk2__default.default.bold("\u{1F4BE} Size Impact\n"));
  console.log(`  Bytes saved:       ${chalk2__default.default.green(formatBytes(metrics.bytesSaved))}`);
  console.log(`  CSS overhead:      ${chalk2__default.default.yellow(formatBytes(metrics.consolidatedCSSBytes))}`);
  console.log(`  Net reduction:     ${chalk2__default.default.green(formatBytes(metrics.bytesSaved - metrics.consolidatedCSSBytes))}`);
  console.log(`  Percentage:        ${chalk2__default.default.green(formatPercentage(metrics.percentageReduction))}`);
  console.log();
  console.log(chalk2__default.default.bold("\u26A1 Browser Impact\n"));
  console.log(`  Parse time saved:  ${chalk2__default.default.green(formatTime(metrics.estimatedParseTimeSavedMs))}`);
  console.log(`  Render time saved: ${chalk2__default.default.green(formatTime(metrics.estimatedRenderTimeSavedMs))}`);
  console.log();
  if (metrics.topConsolidations.length > 0) {
    console.log(chalk2__default.default.bold("\u{1F51D} Top Consolidations\n"));
    for (let i = 0; i < metrics.topConsolidations.length; i++) {
      const c = metrics.topConsolidations[i];
      console.log(
        `  ${chalk2__default.default.gray(`${i + 1}.`)} ${chalk2__default.default.yellow(c.consolidated)} ${chalk2__default.default.gray("(")}${chalk2__default.default.cyan(c.frequency)}x${chalk2__default.default.gray(")")} saves ${chalk2__default.default.green(formatBytes(c.bytesSaved))}`
      );
    }
    console.log();
  }
  console.log(chalk2__default.default.bold("\u{1F4CB} All Mappings\n"));
  for (const mapping of manifest.mappings.slice(0, 20)) {
    console.log(`  ${chalk2__default.default.yellow(mapping.consolidated)} \u2190 ${chalk2__default.default.gray(truncate2(mapping.original, 50))}`);
  }
  if (manifest.mappings.length > 20) {
    console.log(chalk2__default.default.gray(`  ... and ${manifest.mappings.length - 20} more`));
  }
  console.log();
}
function generateHTMLReport(manifest) {
  const { metrics } = manifest;
  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Classpresso Report</title>
  <style>
    body {
      font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, sans-serif;
      max-width: 800px;
      margin: 0 auto;
      padding: 2rem;
      background: #1a1a1a;
      color: #e0e0e0;
    }
    h1 { color: #d4af37; }
    h2 { color: #888; margin-top: 2rem; }
    .metric {
      display: flex;
      justify-content: space-between;
      padding: 0.5rem 0;
      border-bottom: 1px solid #333;
    }
    .metric-value { color: #4ade80; font-weight: bold; }
    .mapping {
      display: flex;
      gap: 1rem;
      padding: 0.5rem;
      background: #222;
      margin: 0.25rem 0;
      border-radius: 4px;
    }
    .hash { color: #d4af37; font-family: monospace; }
    .original { color: #888; font-family: monospace; font-size: 0.9rem; overflow: hidden; text-overflow: ellipsis; }
    table { width: 100%; border-collapse: collapse; margin-top: 1rem; }
    th, td { padding: 0.5rem; text-align: left; border-bottom: 1px solid #333; }
    th { color: #888; }
  </style>
</head>
<body>
  <h1>\u2615 Classpresso Report</h1>
  <p style="color: #888">Generated: ${manifest.created}</p>

  <h2>Summary</h2>
  <div class="metric"><span>Files Scanned</span><span class="metric-value">${metrics.totalFilesScanned}</span></div>
  <div class="metric"><span>Files Modified</span><span class="metric-value">${metrics.totalFilesModified}</span></div>
  <div class="metric"><span>Patterns Consolidated</span><span class="metric-value">${metrics.consolidatedPatterns}</span></div>

  <h2>Size Impact</h2>
  <div class="metric"><span>Bytes Saved</span><span class="metric-value">${formatBytes(metrics.bytesSaved)}</span></div>
  <div class="metric"><span>CSS Overhead</span><span class="metric-value">${formatBytes(metrics.consolidatedCSSBytes)}</span></div>
  <div class="metric"><span>Net Reduction</span><span class="metric-value">${formatBytes(metrics.bytesSaved - metrics.consolidatedCSSBytes)} (${formatPercentage(metrics.percentageReduction)})</span></div>

  <h2>Browser Impact</h2>
  <div class="metric"><span>Parse Time Saved</span><span class="metric-value">${formatTime(metrics.estimatedParseTimeSavedMs)}</span></div>
  <div class="metric"><span>Render Time Saved</span><span class="metric-value">${formatTime(metrics.estimatedRenderTimeSavedMs)}</span></div>

  <h2>Top Consolidations</h2>
  <table>
    <thead>
      <tr><th>#</th><th>Hash</th><th>Frequency</th><th>Savings</th></tr>
    </thead>
    <tbody>
      ${metrics.topConsolidations.map((c, i) => `
        <tr>
          <td>${i + 1}</td>
          <td class="hash">${c.consolidated}</td>
          <td>${c.frequency}x</td>
          <td class="metric-value">${formatBytes(c.bytesSaved)}</td>
        </tr>
      `).join("")}
    </tbody>
  </table>

  <h2>All Mappings</h2>
  ${manifest.mappings.map((m) => `
    <div class="mapping">
      <span class="hash">${m.consolidated}</span>
      <span class="original">${escapeHtml(m.original)}</span>
    </div>
  `).join("")}
</body>
</html>`;
}
function truncate2(str, maxLength) {
  if (str.length <= maxLength) return str;
  return str.substring(0, maxLength - 3) + "...";
}
function escapeHtml(str) {
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}

// src/cli/index.ts
var program = new commander.Command();
program.name("classpresso").description("CSS class consolidation tool - compress multiple utility classes into single optimized classes").version(getVersion());
program.command("analyze").description("Analyze build output and show potential optimizations").option("-d, --dir <path>", "Build directory", ".next").option("--min-occurrences <n>", "Minimum occurrences to consolidate", "2").option("--min-classes <n>", "Minimum classes in pattern", "2").option("--ssr", "Enable SSR-safe mode (only transforms patterns in both HTML and JS)").option("--json", "Output as JSON").option("-v, --verbose", "Verbose output").option("--debug", "Enable debug mode (generates detailed log file)").option("--send-error-reports", "Send error reports to configured webhook").option("--error-report-url <url>", "Webhook URL for error reports").action(analyzeCommand);
program.command("optimize").description("Optimize build output by consolidating classes").option("-d, --dir <path>", "Build directory", ".next").option("--min-occurrences <n>", "Minimum occurrences to consolidate", "2").option("--min-classes <n>", "Minimum classes in pattern", "2").option("--ssr", "Enable SSR-safe mode (only transforms patterns in both HTML and JS)").option("--dry-run", "Show what would be done without making changes").option("--backup", "Create backup before modifying").option("--no-manifest", "Do not generate manifest file").option("--purge-unused", "Remove unused CSS classes after consolidation").option("-v, --verbose", "Verbose output").option("--debug", "Enable debug mode (generates detailed log file)").option("--send-error-reports", "Send error reports to configured webhook").option("--error-report-url <url>", "Webhook URL for error reports").action(optimizeCommand);
program.command("report").description("Generate report from existing manifest").option("-d, --dir <path>", "Build directory", ".next").option("--format <type>", "Output format: text, json, html", "text").action(reportCommand);

exports.program = program;
