// classpresso/dist/consolidator.cjs
'use strict';

var promises = require('fs/promises');
var glob = require('glob');
var path2 = require('path');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var path2__default = /*#__PURE__*/_interopDefault(path2);

var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/utils/files.ts
var files_exports = {};
__export(files_exports, {
  DEFAULT_PATTERNS: () => DEFAULT_PATTERNS,
  backupFile: () => backupFile,
  filterExcludedFiles: () => filterExcludedFiles,
  findFiles: () => findFiles,
  getFileSize: () => getFileSize,
  getFileStats: () => getFileStats,
  isCSSFile: () => isCSSFile,
  isHTMLFile: () => isHTMLFile,
  isJSFile: () => isJSFile,
  isRSCFile: () => isRSCFile,
  readFileContent: () => readFileContent,
  shouldExcludeFile: () => shouldExcludeFile,
  writeFileContent: () => writeFileContent
});
async function findFiles(buildDir, patterns) {
  const globPatterns = [...new Set(patterns.map((p) => path2__default.default.join(buildDir, p)))];
  const allFiles = await glob.glob(globPatterns, {
    nodir: true,
    windowsPathsNoEscape: true
  });
  return [...new Set(allFiles)];
}
async function readFileContent(filePath) {
  return promises.readFile(filePath, "utf-8");
}
async function writeFileContent(filePath, content) {
  await promises.writeFile(filePath, content, "utf-8");
}
async function getFileSize(filePath) {
  const stats = await promises.stat(filePath);
  return stats.size;
}
async function backupFile(filePath) {
  const backupPath = `${filePath}.classpresso-backup`;
  await promises.copyFile(filePath, backupPath);
  return backupPath;
}
async function getFileStats(filePaths) {
  const stats = [];
  for (const filePath of filePaths) {
    try {
      const size = await getFileSize(filePath);
      stats.push({
        path: filePath,
        originalSize: size,
        modified: false
      });
    } catch {
    }
  }
  return stats;
}
function isJSFile(filePath) {
  return /\.(js|mjs|cjs)$/.test(filePath);
}
function isCSSFile(filePath) {
  return /\.css$/.test(filePath);
}
function isHTMLFile(filePath) {
  return /\.(html|htm)$/.test(filePath);
}
function isRSCFile(filePath) {
  return /\.rsc$/.test(filePath);
}
function shouldExcludeFile(filePath, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return false;
  }
  const normalizedPath = filePath.replace(/\\/g, "/");
  for (const pattern of excludePatterns) {
    const regexPattern = pattern.replace(/\\/g, "/").replace(/\*\*/g, "{{DOUBLESTAR}}").replace(/\*/g, "[^/]*").replace(/\?/g, ".").replace(/{{DOUBLESTAR}}/g, ".*");
    const regex = new RegExp(regexPattern);
    if (regex.test(normalizedPath)) {
      return true;
    }
  }
  return false;
}
function filterExcludedFiles(files, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return files;
  }
  return files.filter((filePath) => !shouldExcludeFile(filePath, excludePatterns));
}
var DEFAULT_PATTERNS;
var init_files = __esm({
  "src/utils/files.ts"() {
    DEFAULT_PATTERNS = [
      // === Next.js (.next directory) ===
      // Client-side chunks
      "static/chunks/**/*.js",
      "static/css/**/*.css",
      // Server components (Next.js App Router)
      "server/app/**/*.js",
      "server/app/**/*.html",
      "server/app/**/*.rsc",
      // Server chunks (contains Client Component code used for hydration)
      "server/chunks/**/*.js",
      // Standalone build (Next.js output: 'standalone')
      "standalone/.next/static/chunks/**/*.js",
      "standalone/.next/static/css/**/*.css",
      "standalone/.next/server/app/**/*.js",
      "standalone/.next/server/app/**/*.html",
      "standalone/.next/server/app/**/*.rsc",
      "standalone/.next/server/chunks/**/*.js",
      // === Nuxt (.output directory) ===
      "public/**/*.html",
      "public/**/*.js",
      "public/**/*.css",
      "public/_nuxt/**/*.js",
      "public/_nuxt/**/*.css",
      "server/**/*.mjs",
      "server/chunks/**/*.mjs",
      // === SvelteKit (build directory) ===
      // Static adapter output
      "**/*.html",
      "_app/**/*.js",
      "_app/**/*.css",
      // Node adapter output
      "client/**/*.js",
      "client/**/*.css",
      "server/**/*.js",
      // === Astro (dist directory) ===
      "_astro/**/*.js",
      "_astro/**/*.css",
      // SSR/Hybrid build - Client assets
      "client/_astro/**/*.js",
      "client/_astro/**/*.css",
      // === Remix (build directory) ===
      "client/**/*.js",
      "client/**/*.css",
      "server/**/*.js",
      // === Solid Start (.output or dist directory) ===
      "public/**/*.html",
      "public/**/*.js",
      "public/**/*.css",
      "_build/**/*.js",
      "_build/**/*.css",
      // === Qwik (dist directory) ===
      "build/**/*.js",
      "build/**/*.css",
      "q-*.js",
      // === Angular (dist/[project] directory) ===
      // Angular 17+ uses browser subdirectory
      "browser/**/*.js",
      "browser/**/*.css",
      "browser/**/*.html",
      // Legacy Angular output
      "*.js",
      "*.css",
      "*.html",
      // === Gatsby (public directory) ===
      "page-data/**/*.json",
      "*.js",
      "*.css",
      "*.html",
      // === Vite/Vue/React generic (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "index.html",
      // === Ember (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "*.html",
      // === VitePress (.vitepress/dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === Docusaurus (build directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === Eleventy/11ty (_site directory) ===
      "**/*.html",
      "**/*.js",
      "**/*.css",
      // === Hugo (public directory) ===
      "**/*.html",
      "**/*.js",
      "**/*.css",
      // === Parcel (dist directory) ===
      "*.html",
      "*.js",
      "*.css",
      // === Preact (build directory) ===
      "bundle.js",
      "bundle.*.js",
      "*.css",
      "*.html",
      // === Gridsome (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === RedwoodJS (web/dist directory) ===
      "static/**/*.js",
      "static/**/*.css",
      "*.html",
      // === Webpack generic (dist directory) ===
      "*.js",
      "*.css",
      "*.html",
      "js/**/*.js",
      "css/**/*.css"
    ];
  }
});

// src/core/consolidator.ts
init_files();
function createClassMappings(candidates) {
  return candidates.map((candidate) => ({
    original: candidate.classString,
    consolidated: candidate.hashName,
    classes: candidate.classes,
    excludedClasses: candidate.excludedClasses,
    cssDeclarations: "",
    // Will be filled by CSS generator
    frequency: candidate.frequency,
    bytesSaved: candidate.bytesSaved
  }));
}
function buildReplacementMap(mappings) {
  const map = /* @__PURE__ */ new Map();
  for (const mapping of mappings) {
    const key = [...mapping.classes].sort().join(" ");
    map.set(key, {
      consolidated: mapping.consolidated,
      excludedClasses: mapping.excludedClasses
    });
  }
  return map;
}
async function saveMappingManifest(mappings, metrics, config) {
  const manifest = {
    version: "1.0.0",
    tool: "classpresso",
    buildDir: config.buildDir,
    created: (/* @__PURE__ */ new Date()).toISOString(),
    config: {
      minOccurrences: config.minOccurrences,
      minClasses: config.minClasses,
      hashPrefix: config.hashPrefix,
      hashLength: config.hashLength
    },
    mappings,
    metrics
  };
  const manifestPath = path2__default.default.join(config.buildDir, "classpresso-manifest.json");
  await writeFileContent(manifestPath, JSON.stringify(manifest, null, 2));
  return manifestPath;
}
async function loadMappingManifest(buildDir) {
  const manifestPath = path2__default.default.join(buildDir, "classpresso-manifest.json");
  try {
    const { readFileContent: readFileContent2 } = await Promise.resolve().then(() => (init_files(), files_exports));
    const content = await readFileContent2(manifestPath);
    return JSON.parse(content);
  } catch {
    return null;
  }
}

exports.buildReplacementMap = buildReplacementMap;
exports.createClassMappings = createClassMappings;
exports.loadMappingManifest = loadMappingManifest;
exports.saveMappingManifest = saveMappingManifest;
