// classpresso/dist/index.js
import { readFile, writeFile, stat, copyFile } from 'fs/promises';
import { glob } from 'glob';
import path2 from 'path';
import crypto from 'crypto';
import postcss from 'postcss';

var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/utils/files.ts
var files_exports = {};
__export(files_exports, {
  DEFAULT_PATTERNS: () => DEFAULT_PATTERNS,
  backupFile: () => backupFile,
  filterExcludedFiles: () => filterExcludedFiles,
  findFiles: () => findFiles,
  getFileSize: () => getFileSize,
  getFileStats: () => getFileStats,
  isCSSFile: () => isCSSFile,
  isHTMLFile: () => isHTMLFile,
  isJSFile: () => isJSFile,
  isRSCFile: () => isRSCFile,
  readFileContent: () => readFileContent,
  shouldExcludeFile: () => shouldExcludeFile,
  writeFileContent: () => writeFileContent
});
async function findFiles(buildDir, patterns) {
  const globPatterns = [...new Set(patterns.map((p) => path2.join(buildDir, p)))];
  const allFiles = await glob(globPatterns, {
    nodir: true,
    windowsPathsNoEscape: true
  });
  return [...new Set(allFiles)];
}
async function readFileContent(filePath) {
  return readFile(filePath, "utf-8");
}
async function writeFileContent(filePath, content) {
  await writeFile(filePath, content, "utf-8");
}
async function getFileSize(filePath) {
  const stats = await stat(filePath);
  return stats.size;
}
async function backupFile(filePath) {
  const backupPath = `${filePath}.classpresso-backup`;
  await copyFile(filePath, backupPath);
  return backupPath;
}
async function getFileStats(filePaths) {
  const stats = [];
  for (const filePath of filePaths) {
    try {
      const size = await getFileSize(filePath);
      stats.push({
        path: filePath,
        originalSize: size,
        modified: false
      });
    } catch {
    }
  }
  return stats;
}
function isJSFile(filePath) {
  return /\.(js|mjs|cjs)$/.test(filePath);
}
function isCSSFile(filePath) {
  return /\.css$/.test(filePath);
}
function isHTMLFile(filePath) {
  return /\.(html|htm)$/.test(filePath);
}
function isRSCFile(filePath) {
  return /\.rsc$/.test(filePath);
}
function shouldExcludeFile(filePath, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return false;
  }
  const normalizedPath = filePath.replace(/\\/g, "/");
  for (const pattern of excludePatterns) {
    const regexPattern = pattern.replace(/\\/g, "/").replace(/\*\*/g, "{{DOUBLESTAR}}").replace(/\*/g, "[^/]*").replace(/\?/g, ".").replace(/{{DOUBLESTAR}}/g, ".*");
    const regex = new RegExp(regexPattern);
    if (regex.test(normalizedPath)) {
      return true;
    }
  }
  return false;
}
function filterExcludedFiles(files, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return files;
  }
  return files.filter((filePath) => !shouldExcludeFile(filePath, excludePatterns));
}
var DEFAULT_PATTERNS;
var init_files = __esm({
  "src/utils/files.ts"() {
    DEFAULT_PATTERNS = [
      // === Next.js (.next directory) ===
      // Client-side chunks
      "static/chunks/**/*.js",
      "static/css/**/*.css",
      // Server components (Next.js App Router)
      "server/app/**/*.js",
      "server/app/**/*.html",
      "server/app/**/*.rsc",
      // Server chunks (contains Client Component code used for hydration)
      "server/chunks/**/*.js",
      // Standalone build (Next.js output: 'standalone')
      "standalone/.next/static/chunks/**/*.js",
      "standalone/.next/static/css/**/*.css",
      "standalone/.next/server/app/**/*.js",
      "standalone/.next/server/app/**/*.html",
      "standalone/.next/server/app/**/*.rsc",
      "standalone/.next/server/chunks/**/*.js",
      // === Nuxt (.output directory) ===
      "public/**/*.html",
      "public/**/*.js",
      "public/**/*.css",
      "public/_nuxt/**/*.js",
      "public/_nuxt/**/*.css",
      "server/**/*.mjs",
      "server/chunks/**/*.mjs",
      // === SvelteKit (build directory) ===
      // Static adapter output
      "**/*.html",
      "_app/**/*.js",
      "_app/**/*.css",
      // Node adapter output
      "client/**/*.js",
      "client/**/*.css",
      "server/**/*.js",
      // === Astro (dist directory) ===
      "_astro/**/*.js",
      "_astro/**/*.css",
      // SSR/Hybrid build - Client assets
      "client/_astro/**/*.js",
      "client/_astro/**/*.css",
      // === Remix (build directory) ===
      "client/**/*.js",
      "client/**/*.css",
      "server/**/*.js",
      // === Solid Start (.output or dist directory) ===
      "public/**/*.html",
      "public/**/*.js",
      "public/**/*.css",
      "_build/**/*.js",
      "_build/**/*.css",
      // === Qwik (dist directory) ===
      "build/**/*.js",
      "build/**/*.css",
      "q-*.js",
      // === Angular (dist/[project] directory) ===
      // Angular 17+ uses browser subdirectory
      "browser/**/*.js",
      "browser/**/*.css",
      "browser/**/*.html",
      // Legacy Angular output
      "*.js",
      "*.css",
      "*.html",
      // === Gatsby (public directory) ===
      "page-data/**/*.json",
      "*.js",
      "*.css",
      "*.html",
      // === Vite/Vue/React generic (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "index.html",
      // === Ember (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "*.html",
      // === VitePress (.vitepress/dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === Docusaurus (build directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === Eleventy/11ty (_site directory) ===
      "**/*.html",
      "**/*.js",
      "**/*.css",
      // === Hugo (public directory) ===
      "**/*.html",
      "**/*.js",
      "**/*.css",
      // === Parcel (dist directory) ===
      "*.html",
      "*.js",
      "*.css",
      // === Preact (build directory) ===
      "bundle.js",
      "bundle.*.js",
      "*.css",
      "*.html",
      // === Gridsome (dist directory) ===
      "assets/**/*.js",
      "assets/**/*.css",
      "**/*.html",
      // === RedwoodJS (web/dist directory) ===
      "static/**/*.js",
      "static/**/*.css",
      "*.html",
      // === Webpack generic (dist directory) ===
      "*.js",
      "*.css",
      "*.html",
      "js/**/*.js",
      "css/**/*.css"
    ];
  }
});

// src/core/scanner.ts
init_files();

// src/utils/regex.ts
function escapeRegex(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
var CLASS_PATTERNS = {
  // JSX: className="..."
  jsxDouble: /className\s*=\s*"([^"]+)"/g,
  // JSX: className='...'
  jsxSingle: /className\s*=\s*'([^']+)'/g,
  // React createElement: className:"..."
  createElementDouble: /className\s*:\s*"([^"]+)"/g,
  // React createElement: className:'...'
  createElementSingle: /className\s*:\s*'([^']+)'/g,
  // Minified: "className","..."
  minifiedComma: /"className"\s*,\s*"([^"]+)"/g,
  // HTML: class="..."
  htmlDouble: /\bclass\s*=\s*"([^"]+)"/g,
  // HTML: class='...'
  htmlSingle: /\bclass\s*=\s*'([^']+)'/g,
  // HTML entity encoded: class=&quot;...&quot;
  htmlEntityDouble: /\bclass=&quot;([^&]+)&quot;/g,
  // HTML entity encoded: class=&#34;...&#34; (numeric entity)
  htmlNumericDouble: /\bclass=&#34;([^&]+)&#34;/g,
  // RSC payload: \"className\":\"...\" (escaped JSON in Next.js RSC payloads)
  rscPayload: /\\"className\\"\s*:\s*\\"([^"\\]+)\\"/g,
  // Vue/Quasar createElement: class:"..." (Vue uses 'class' not 'className')
  vueCreateElementDouble: /\bclass\s*:\s*"([^"]+)"/g,
  // Vue/Quasar createElement: class:'...'
  vueCreateElementSingle: /\bclass\s*:\s*'([^']+)'/g
};
var ALL_CLASS_PATTERNS = [
  CLASS_PATTERNS.jsxDouble,
  CLASS_PATTERNS.jsxSingle,
  CLASS_PATTERNS.createElementDouble,
  CLASS_PATTERNS.createElementSingle,
  CLASS_PATTERNS.minifiedComma,
  CLASS_PATTERNS.htmlDouble,
  CLASS_PATTERNS.htmlSingle,
  CLASS_PATTERNS.htmlEntityDouble,
  CLASS_PATTERNS.htmlNumericDouble,
  CLASS_PATTERNS.rscPayload,
  CLASS_PATTERNS.vueCreateElementDouble,
  CLASS_PATTERNS.vueCreateElementSingle
];
function isDynamicClassString(classString) {
  if (classString.includes("${")) return true;
  if (/\s*\?\s*/.test(classString) && classString.includes(":")) return true;
  if (/\w+\s*\(/.test(classString)) return true;
  if (/^[a-z][a-zA-Z0-9]*$/.test(classString) && !classString.includes("-")) return true;
  return false;
}
var DYNAMIC_BASE_PATTERNS = [
  // className:`base classes ${...}`
  /className\s*:\s*`([^`$]+)\s*\$\{/g,
  // className=`base classes ${...}` (JSX)
  /className\s*=\s*`([^`$]+)\s*\$\{/g,
  // class:`base classes ${...}` (rare but possible)
  /\bclass\s*:\s*`([^`$]+)\s*\$\{/g
];
var CONCAT_BASE_PATTERNS = [
  // className:"".concat("base classes",...) - the base before concat
  /className\s*:\s*""\s*\.concat\s*\(\s*"([^"]+)"/g,
  // className:''.concat('base classes',...)
  /className\s*:\s*''\s*\.concat\s*\(\s*'([^']+)'/g
];
function extractConcatBaseStrings(content) {
  const results = [];
  for (const pattern of CONCAT_BASE_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const baseString = match[1]?.trim();
      if (baseString && baseString.length > 0) {
        results.push(baseString);
      }
    }
  }
  return results;
}
function extractDynamicBaseStrings(content) {
  const results = [];
  for (const pattern of DYNAMIC_BASE_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const baseString = match[1]?.trim();
      if (baseString && baseString.length > 0) {
        results.push(baseString);
      }
    }
  }
  return results;
}

// src/core/scanner.ts
function getSourceFileType(filePath) {
  if (isHTMLFile(filePath)) return "html";
  if (isRSCFile(filePath)) return "rsc";
  return "js";
}
function isProperSubset(classesA, classesB) {
  if (classesA.length >= classesB.length) return false;
  const setB = new Set(classesB);
  return classesA.every((cls) => setB.has(cls));
}
function detectMergeablePatterns(occurrences) {
  const mergeablePatterns = /* @__PURE__ */ new Set();
  const jsPatterns = [];
  const htmlPatterns = [];
  for (const [, occurrence] of occurrences) {
    if (occurrence.sourceTypes.has("js")) {
      jsPatterns.push(occurrence);
    }
    if (occurrence.sourceTypes.has("html") || occurrence.sourceTypes.has("rsc")) {
      htmlPatterns.push(occurrence);
    }
  }
  for (const jsPattern of jsPatterns) {
    for (const htmlPattern of htmlPatterns) {
      if (isProperSubset(jsPattern.classes, htmlPattern.classes)) {
        mergeablePatterns.add(jsPattern.normalizedKey);
        break;
      }
    }
  }
  return mergeablePatterns;
}
function containsDynamicPrefix(classes, dynamicPrefixes) {
  for (const cls of classes) {
    for (const prefix of dynamicPrefixes) {
      if (prefix.endsWith("-")) {
        if (cls.startsWith(prefix)) {
          return true;
        }
      } else {
        if (cls === prefix || cls.startsWith(prefix + "-")) {
          return true;
        }
      }
    }
  }
  return false;
}
function isHydrationSafe(occurrence) {
  const hasServerContext = occurrence.sourceTypes.has("html") || occurrence.sourceTypes.has("rsc");
  const hasClientContext = occurrence.sourceTypes.has("js");
  return hasServerContext && hasClientContext;
}
function shouldExcludeClass(className, exclude) {
  if (exclude.prefixes) {
    for (const prefix of exclude.prefixes) {
      if (className.startsWith(prefix)) return true;
    }
  }
  if (exclude.suffixes) {
    for (const suffix of exclude.suffixes) {
      if (className.endsWith(suffix)) return true;
    }
  }
  if (exclude.classes) {
    if (exclude.classes.includes(className)) return true;
  }
  if (exclude.patterns) {
    for (const pattern of exclude.patterns) {
      if (pattern.test(className)) return true;
    }
  }
  return false;
}
function normalizeClassString(classString, exclude) {
  const allClasses = classString.split(/\s+/).filter(Boolean);
  const includedClasses = [];
  const excludedClasses = [];
  for (const cls of allClasses) {
    if (shouldExcludeClass(cls, exclude)) {
      excludedClasses.push(cls);
    } else {
      includedClasses.push(cls);
    }
  }
  const sorted = [...includedClasses].sort();
  return {
    normalized: sorted.join(" "),
    classes: includedClasses,
    excludedClasses
  };
}
function extractClassStrings(content, filePath) {
  const results = [];
  const seen = /* @__PURE__ */ new Set();
  for (const pattern of ALL_CLASS_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const classString = match[1];
      if (!classString || !classString.trim()) continue;
      if (isDynamicClassString(classString)) continue;
      const key = `${filePath}:${match.index}:${classString}`;
      if (seen.has(key)) continue;
      seen.add(key);
      const beforeMatch = content.substring(0, match.index);
      const line = (beforeMatch.match(/\n/g) || []).length + 1;
      results.push({
        classString: classString.trim(),
        location: {
          filePath,
          line
        }
      });
    }
  }
  return results;
}
async function scanBuildOutput(config) {
  const occurrences = /* @__PURE__ */ new Map();
  const dynamicBasePatterns = /* @__PURE__ */ new Map();
  const fileStats = [];
  const errors = [];
  const patterns = config.include.length > 0 ? config.include : DEFAULT_PATTERNS;
  const allFiles = await findFiles(config.buildDir, patterns);
  const files = filterExcludedFiles(allFiles, config.exclude.files || []);
  if (config.verbose) {
    console.log(`Found ${allFiles.length} files, scanning ${files.length} (${allFiles.length - files.length} excluded)`);
  }
  for (const filePath of files) {
    try {
      if (!isJSFile(filePath) && !isHTMLFile(filePath) && !isRSCFile(filePath)) {
        continue;
      }
      const content = await readFileContent(filePath);
      const size = await getFileSize(filePath);
      fileStats.push({
        path: filePath,
        originalSize: size,
        modified: false
      });
      if (isJSFile(filePath)) {
        const dynamicBases = extractDynamicBaseStrings(content);
        const concatBases = extractConcatBaseStrings(content);
        const allDynamicBases = [...dynamicBases, ...concatBases];
        for (const baseString of allDynamicBases) {
          const { normalized, classes } = normalizeClassString(baseString, config.exclude);
          if (classes.length < config.minClasses) continue;
          const location = { filePath };
          if (dynamicBasePatterns.has(normalized)) {
            dynamicBasePatterns.get(normalized).locations.push(location);
          } else {
            dynamicBasePatterns.set(normalized, {
              baseClasses: classes,
              normalizedKey: normalized,
              locations: [location]
            });
          }
        }
      }
      const classStrings = extractClassStrings(content, filePath);
      const sourceType = getSourceFileType(filePath);
      for (const { classString, location } of classStrings) {
        const { normalized, classes, excludedClasses } = normalizeClassString(
          classString,
          config.exclude
        );
        if (classes.length === 0) continue;
        if (classes.length < config.minClasses) continue;
        if (config.excludeDynamicPatterns && containsDynamicPrefix(classes, config.dynamicPrefixes)) {
          continue;
        }
        if (occurrences.has(normalized)) {
          const existing = occurrences.get(normalized);
          existing.count++;
          existing.locations.push(location);
          existing.sourceTypes.add(sourceType);
        } else {
          occurrences.set(normalized, {
            classString,
            normalizedKey: normalized,
            count: 1,
            locations: [location],
            classes,
            excludedClasses,
            sourceTypes: /* @__PURE__ */ new Set([sourceType])
          });
        }
      }
    } catch (err) {
      errors.push(`Error scanning ${filePath}: ${err}`);
    }
  }
  const mergeablePatterns = detectMergeablePatterns(occurrences);
  if (config.verbose && mergeablePatterns.size > 0) {
    console.log(`Found ${mergeablePatterns.size} patterns that appear to be merged className props`);
  }
  return {
    occurrences,
    files: fileStats,
    errors,
    dynamicBasePatterns,
    mergeablePatterns
  };
}
var ALPHABET = "abcdefghijklmnopqrstuvwxyz0123456789";
var BASE = ALPHABET.length;
function generateSequentialName(index, prefix = "cp-") {
  let result = "";
  let n = index;
  do {
    result = ALPHABET[n % BASE] + result;
    n = Math.floor(n / BASE) - 1;
  } while (n >= 0);
  return `${prefix}${result}`;
}
function generateHashName(normalizedClassString, prefix = "cp-", length = 5) {
  const hash = crypto.createHash("md5").update(normalizedClassString).digest("hex").substring(0, length);
  return `${prefix}${hash}`;
}
function resolveCollisions(candidates) {
  const hashMap = /* @__PURE__ */ new Map();
  for (const candidate of candidates) {
    let hash = candidate.hashName;
    let suffix = 0;
    while (hashMap.has(hash) && hashMap.get(hash).normalizedKey !== candidate.normalizedKey) {
      suffix++;
      hash = `${candidate.hashName}${suffix}`;
    }
    candidate.hashName = hash;
    hashMap.set(hash, candidate);
  }
  return candidates;
}

// src/core/metrics.ts
function calculateMetrics(candidates, originalFiles, transformResult, cssOverhead) {
  const totalBytesSaved = candidates.reduce((sum, c) => sum + c.bytesSaved, 0);
  const netSavings = totalBytesSaved - cssOverhead;
  const originalTotalBytes = originalFiles.reduce((sum, f) => sum + f.originalSize, 0);
  const optimizedTotalBytes = originalTotalBytes - netSavings;
  const estimatedParseTimeSavedMs = Math.max(0, netSavings / 1024 * 0.1);
  const estimatedRenderTimeSavedMs = estimatedParseTimeSavedMs * 0.5;
  return {
    totalFilesScanned: originalFiles.length,
    totalFilesModified: transformResult.filesModified,
    totalClassStringsFound: candidates.reduce((sum, c) => sum + c.frequency, 0),
    uniqueClassPatterns: candidates.length,
    consolidatedPatterns: candidates.filter((c) => c.frequency >= 2).length,
    totalOccurrencesReplaced: candidates.reduce((sum, c) => sum + c.frequency, 0),
    originalTotalBytes,
    optimizedTotalBytes,
    bytesSaved: totalBytesSaved,
    percentageReduction: originalTotalBytes > 0 ? netSavings / originalTotalBytes * 100 : 0,
    originalCSSBytes: 0,
    // Would need to track CSS separately
    consolidatedCSSBytes: cssOverhead,
    netCSSChange: cssOverhead,
    estimatedParseTimeSavedMs,
    estimatedRenderTimeSavedMs,
    topConsolidations: candidates.slice(0, 10).map((c) => ({
      original: c.classString,
      consolidated: c.hashName,
      frequency: c.frequency,
      bytesSaved: c.bytesSaved
    }))
  };
}
function calculatePatternCSSOverhead(classes) {
  const baseOverhead = 15;
  const perClassOverhead = 20;
  return baseOverhead + classes.length * perClassOverhead;
}
function estimateCSSOverhead(candidates) {
  return candidates.reduce((sum, c) => {
    return sum + calculatePatternCSSOverhead(c.classes);
  }, 0);
}
function formatBytes(bytes) {
  if (bytes === 0) return "0 B";
  const sign = bytes < 0 ? "-" : "";
  const absBytes = Math.abs(bytes);
  if (absBytes < 1024) {
    return `${sign}${absBytes} B`;
  } else if (absBytes < 1024 * 1024) {
    return `${sign}${(absBytes / 1024).toFixed(2)} KB`;
  } else {
    return `${sign}${(absBytes / (1024 * 1024)).toFixed(2)} MB`;
  }
}
function formatPercentage(value) {
  return `${value.toFixed(2)}%`;
}
function formatTime(ms) {
  if (ms < 1) {
    return `${(ms * 1e3).toFixed(2)}\u03BCs`;
  } else if (ms < 1e3) {
    return `${ms.toFixed(2)}ms`;
  } else {
    return `${(ms / 1e3).toFixed(2)}s`;
  }
}

// src/core/pattern-detector.ts
function calculateBytesSaved(original, hashName, frequency, excludedClasses) {
  const originalLength = original.length;
  const excludedLength = excludedClasses.join(" ").length;
  const includedLength = originalLength - excludedLength - (excludedClasses.length > 0 ? 1 : 0);
  const newLength = hashName.length;
  const savingsPerOccurrence = includedLength - newLength;
  return savingsPerOccurrence * frequency;
}
function detectConsolidatablePatterns(occurrences, config, mergeablePatterns) {
  const preliminaryCandidates = [];
  for (const [, occurrence] of occurrences) {
    if (occurrence.count < config.minOccurrences) continue;
    if (occurrence.classes.length < config.minClasses) continue;
    if (config.ssr && !isHydrationSafe(occurrence)) {
      continue;
    }
    if (config.ssr && mergeablePatterns && mergeablePatterns.has(occurrence.normalizedKey)) {
      continue;
    }
    if (config.skipPatternsWithExcludedClasses && occurrence.excludedClasses.length > 0) {
      continue;
    }
    const cssOverhead = calculatePatternCSSOverhead(occurrence.classes);
    preliminaryCandidates.push({ occurrence, cssOverhead });
  }
  preliminaryCandidates.sort((a, b) => b.occurrence.count - a.occurrence.count);
  const filteredCandidates = [];
  for (const { occurrence, cssOverhead } of preliminaryCandidates) {
    const estimatedNameLength = config.hashPrefix.length + 1;
    const estimatedBytesSaved = calculateBytesSaved(
      occurrence.classString,
      "x".repeat(estimatedNameLength),
      // placeholder
      occurrence.count,
      occurrence.excludedClasses
    );
    if (estimatedBytesSaved < config.minBytesSaved) continue;
    if (!config.forceAll) {
      if (estimatedBytesSaved <= cssOverhead) continue;
    }
    filteredCandidates.push({ occurrence, cssOverhead });
  }
  const candidates = [];
  for (let i = 0; i < filteredCandidates.length; i++) {
    const { occurrence } = filteredCandidates[i];
    const hashName = generateSequentialName(i, config.hashPrefix);
    const bytesSaved = calculateBytesSaved(
      occurrence.classString,
      hashName,
      occurrence.count,
      occurrence.excludedClasses
    );
    candidates.push({
      classString: occurrence.classString,
      normalizedKey: occurrence.normalizedKey,
      frequency: occurrence.count,
      bytesSaved,
      classes: occurrence.classes,
      excludedClasses: occurrence.excludedClasses,
      hashName
    });
  }
  return candidates.sort((a, b) => b.bytesSaved - a.bytesSaved);
}
function getPatternSummary(candidates) {
  const totalPatterns = candidates.length;
  const totalOccurrences = candidates.reduce((sum, c) => sum + c.frequency, 0);
  const totalBytesSaved = candidates.reduce((sum, c) => sum + c.bytesSaved, 0);
  const totalClasses = candidates.reduce((sum, c) => sum + c.classes.length, 0);
  return {
    totalPatterns,
    totalOccurrences,
    totalBytesSaved,
    avgFrequency: totalPatterns > 0 ? totalOccurrences / totalPatterns : 0,
    avgClassesPerPattern: totalPatterns > 0 ? totalClasses / totalPatterns : 0
  };
}

// src/core/consolidator.ts
init_files();
function createClassMappings(candidates) {
  return candidates.map((candidate) => ({
    original: candidate.classString,
    consolidated: candidate.hashName,
    classes: candidate.classes,
    excludedClasses: candidate.excludedClasses,
    cssDeclarations: "",
    // Will be filled by CSS generator
    frequency: candidate.frequency,
    bytesSaved: candidate.bytesSaved
  }));
}
function buildReplacementMap(mappings) {
  const map = /* @__PURE__ */ new Map();
  for (const mapping of mappings) {
    const key = [...mapping.classes].sort().join(" ");
    map.set(key, {
      consolidated: mapping.consolidated,
      excludedClasses: mapping.excludedClasses
    });
  }
  return map;
}
async function saveMappingManifest(mappings, metrics, config) {
  const manifest = {
    version: "1.0.0",
    tool: "classpresso",
    buildDir: config.buildDir,
    created: (/* @__PURE__ */ new Date()).toISOString(),
    config: {
      minOccurrences: config.minOccurrences,
      minClasses: config.minClasses,
      hashPrefix: config.hashPrefix,
      hashLength: config.hashLength
    },
    mappings,
    metrics
  };
  const manifestPath = path2.join(config.buildDir, "classpresso-manifest.json");
  await writeFileContent(manifestPath, JSON.stringify(manifest, null, 2));
  return manifestPath;
}
async function loadMappingManifest(buildDir) {
  const manifestPath = path2.join(buildDir, "classpresso-manifest.json");
  try {
    const { readFileContent: readFileContent2 } = await Promise.resolve().then(() => (init_files(), files_exports));
    const content = await readFileContent2(manifestPath);
    return JSON.parse(content);
  } catch {
    return null;
  }
}

// src/core/css-generator.ts
init_files();
var UTILITY_PROPERTY_MAP = {
  // Display
  "flex": "display: flex",
  "grid": "display: grid",
  "block": "display: block",
  "inline": "display: inline",
  "inline-block": "display: inline-block",
  "hidden": "display: none",
  // Flex direction
  "flex-col": "flex-direction: column",
  "flex-row": "flex-direction: row",
  "flex-col-reverse": "flex-direction: column-reverse",
  "flex-row-reverse": "flex-direction: row-reverse",
  // Flex wrap
  "flex-wrap": "flex-wrap: wrap",
  "flex-nowrap": "flex-wrap: nowrap",
  // Justify content
  "justify-start": "justify-content: flex-start",
  "justify-end": "justify-content: flex-end",
  "justify-center": "justify-content: center",
  "justify-between": "justify-content: space-between",
  "justify-around": "justify-content: space-around",
  "justify-evenly": "justify-content: space-evenly",
  // Align items
  "items-start": "align-items: flex-start",
  "items-end": "align-items: flex-end",
  "items-center": "align-items: center",
  "items-baseline": "align-items: baseline",
  "items-stretch": "align-items: stretch",
  // Text align
  "text-left": "text-align: left",
  "text-center": "text-align: center",
  "text-right": "text-align: right",
  // Font weight
  "font-bold": "font-weight: 700",
  "font-semibold": "font-weight: 600",
  "font-medium": "font-weight: 500",
  "font-normal": "font-weight: 400",
  "font-light": "font-weight: 300",
  // Position
  "relative": "position: relative",
  "absolute": "position: absolute",
  "fixed": "position: fixed",
  "sticky": "position: sticky",
  // Overflow
  "overflow-hidden": "overflow: hidden",
  "overflow-auto": "overflow: auto",
  "overflow-scroll": "overflow: scroll",
  "overflow-visible": "overflow: visible",
  // Cursor
  "cursor-pointer": "cursor: pointer",
  "cursor-default": "cursor: default",
  "cursor-not-allowed": "cursor: not-allowed",
  // Misc
  "truncate": ["overflow: hidden", "text-overflow: ellipsis", "white-space: nowrap"],
  "uppercase": "text-transform: uppercase",
  "lowercase": "text-transform: lowercase",
  "capitalize": "text-transform: capitalize"
};
function parseUtilityClass(className) {
  if (UTILITY_PROPERTY_MAP[className]) {
    const value = UTILITY_PROPERTY_MAP[className];
    return Array.isArray(value) ? value : [value];
  }
  const arbitraryMatch = className.match(/^([a-z]+(?:-[a-z]+)?)-\[(.+)\]$/);
  if (arbitraryMatch) {
    const [, prefix, value] = arbitraryMatch;
    return parseArbitraryValue(prefix, value);
  }
  const spacingMatch = className.match(/^(p|px|py|pt|pb|pl|pr|m|mx|my|mt|mb|ml|mr|gap|gap-x|gap-y)-(\d+(?:\.\d+)?|\[.+\])$/);
  if (spacingMatch) {
    const [, prefix, value] = spacingMatch;
    return parseSpacingUtility(prefix, value);
  }
  const sizeMatch = className.match(/^(w|h|min-w|min-h|max-w|max-h)-(.+)$/);
  if (sizeMatch) {
    const [, prefix, value] = sizeMatch;
    return parseSizeUtility(prefix, value);
  }
  const roundedMatch = className.match(/^rounded(-[a-z]+)?(-[a-z0-9]+)?$/);
  if (roundedMatch) {
    return parseRoundedUtility(className);
  }
  const textSizeMatch = className.match(/^text-(xs|sm|base|lg|xl|2xl|3xl|4xl|5xl|6xl|7xl|8xl|9xl)$/);
  if (textSizeMatch) {
    return parseTextSizeUtility(textSizeMatch[1]);
  }
  const opacityMatch = className.match(/^opacity-(\d+)$/);
  if (opacityMatch) {
    return [`opacity: ${parseInt(opacityMatch[1]) / 100}`];
  }
  const zMatch = className.match(/^z-(\d+|auto)$/);
  if (zMatch) {
    return [`z-index: ${zMatch[1]}`];
  }
  return [];
}
function parseArbitraryValue(prefix, value) {
  const propertyMap = {
    "text": "color",
    "bg": "background-color",
    "border": "border-color",
    "border-t": "border-top-color",
    "border-b": "border-bottom-color",
    "border-l": "border-left-color",
    "border-r": "border-right-color",
    "fill": "fill",
    "stroke": "stroke",
    "p": "padding",
    "px": "padding-inline",
    "py": "padding-block",
    "pt": "padding-top",
    "pb": "padding-bottom",
    "pl": "padding-left",
    "pr": "padding-right",
    "m": "margin",
    "mx": "margin-inline",
    "my": "margin-block",
    "mt": "margin-top",
    "mb": "margin-bottom",
    "ml": "margin-left",
    "mr": "margin-right",
    "w": "width",
    "h": "height",
    "min-w": "min-width",
    "min-h": "min-height",
    "max-w": "max-width",
    "max-h": "max-height",
    "top": "top",
    "bottom": "bottom",
    "left": "left",
    "right": "right",
    "gap": "gap",
    "gap-x": "column-gap",
    "gap-y": "row-gap"
  };
  const property = propertyMap[prefix];
  if (property) {
    return [`${property}: ${value}`];
  }
  return [];
}
function parseSpacingUtility(prefix, value) {
  const spacingScale = {
    "0": "0px",
    "0.5": "0.125rem",
    "1": "0.25rem",
    "1.5": "0.375rem",
    "2": "0.5rem",
    "2.5": "0.625rem",
    "3": "0.75rem",
    "3.5": "0.875rem",
    "4": "1rem",
    "5": "1.25rem",
    "6": "1.5rem",
    "7": "1.75rem",
    "8": "2rem",
    "9": "2.25rem",
    "10": "2.5rem",
    "11": "2.75rem",
    "12": "3rem",
    "14": "3.5rem",
    "16": "4rem",
    "20": "5rem",
    "24": "6rem",
    "28": "7rem",
    "32": "8rem",
    "36": "9rem",
    "40": "10rem",
    "44": "11rem",
    "48": "12rem",
    "52": "13rem",
    "56": "14rem",
    "60": "15rem",
    "64": "16rem",
    "72": "18rem",
    "80": "20rem",
    "96": "24rem"
  };
  let cssValue = spacingScale[value];
  if (!cssValue && value.startsWith("[") && value.endsWith("]")) {
    cssValue = value.slice(1, -1);
  }
  if (!cssValue) return [];
  const propertyMap = {
    "p": "padding",
    "px": ["padding-left", "padding-right"],
    "py": ["padding-top", "padding-bottom"],
    "pt": "padding-top",
    "pb": "padding-bottom",
    "pl": "padding-left",
    "pr": "padding-right",
    "m": "margin",
    "mx": ["margin-left", "margin-right"],
    "my": ["margin-top", "margin-bottom"],
    "mt": "margin-top",
    "mb": "margin-bottom",
    "ml": "margin-left",
    "mr": "margin-right",
    "gap": "gap",
    "gap-x": "column-gap",
    "gap-y": "row-gap"
  };
  const property = propertyMap[prefix];
  if (Array.isArray(property)) {
    return property.map((p) => `${p}: ${cssValue}`);
  }
  if (property) {
    return [`${property}: ${cssValue}`];
  }
  return [];
}
function parseSizeUtility(prefix, value) {
  const sizeMap = {
    "full": "100%",
    "screen": "100vh",
    "svh": "100svh",
    "lvh": "100lvh",
    "dvh": "100dvh",
    "min": "min-content",
    "max": "max-content",
    "fit": "fit-content",
    "auto": "auto"
  };
  let cssValue = sizeMap[value];
  if (!cssValue && value.startsWith("[") && value.endsWith("]")) {
    cssValue = value.slice(1, -1);
  }
  if (!cssValue) {
    const spacingMatch = value.match(/^(\d+(?:\.\d+)?)$/);
    if (spacingMatch) {
      cssValue = `${parseFloat(spacingMatch[1]) * 0.25}rem`;
    }
  }
  if (!cssValue) return [];
  const propertyMap = {
    "w": "width",
    "h": "height",
    "min-w": "min-width",
    "min-h": "min-height",
    "max-w": "max-width",
    "max-h": "max-height"
  };
  const property = propertyMap[prefix];
  if (property) {
    return [`${property}: ${cssValue}`];
  }
  return [];
}
function parseRoundedUtility(className) {
  const roundedMap = {
    "rounded": "border-radius: 0.25rem",
    "rounded-none": "border-radius: 0px",
    "rounded-sm": "border-radius: 0.125rem",
    "rounded-md": "border-radius: 0.375rem",
    "rounded-lg": "border-radius: 0.5rem",
    "rounded-xl": "border-radius: 0.75rem",
    "rounded-2xl": "border-radius: 1rem",
    "rounded-3xl": "border-radius: 1.5rem",
    "rounded-full": "border-radius: 9999px"
  };
  if (roundedMap[className]) {
    return [roundedMap[className]];
  }
  return [];
}
function parseTextSizeUtility(size) {
  const sizeMap = {
    "xs": ["0.75rem", "1rem"],
    "sm": ["0.875rem", "1.25rem"],
    "base": ["1rem", "1.5rem"],
    "lg": ["1.125rem", "1.75rem"],
    "xl": ["1.25rem", "1.75rem"],
    "2xl": ["1.5rem", "2rem"],
    "3xl": ["1.875rem", "2.25rem"],
    "4xl": ["2.25rem", "2.5rem"],
    "5xl": ["3rem", "1"],
    "6xl": ["3.75rem", "1"],
    "7xl": ["4.5rem", "1"],
    "8xl": ["6rem", "1"],
    "9xl": ["8rem", "1"]
  };
  const [fontSize, lineHeight] = sizeMap[size] || ["1rem", "1.5rem"];
  return [`font-size: ${fontSize}`, `line-height: ${lineHeight}`];
}
async function extractUtilityCSS(buildDir) {
  const utilityMap = /* @__PURE__ */ new Map();
  const cssPatterns = [
    // Next.js
    "static/**/*.css",
    "static/css/**/*.css",
    // Angular
    "browser/**/*.css",
    // Astro
    "_astro/**/*.css",
    // Nuxt
    "public/_nuxt/**/*.css",
    // Vite/Vue/React/Qwik generic
    "assets/**/*.css",
    // Gatsby/Hugo/Eleventy/Docusaurus/Parcel
    "**/*.css"
  ];
  const cssFiles = await findFiles(buildDir, cssPatterns);
  for (const cssFile of cssFiles) {
    if (!isCSSFile(cssFile)) continue;
    try {
      const content = await readFileContent(cssFile);
      const root = postcss.parse(content);
      root.walkRules((rule) => {
        const selectorMatch = rule.selector.match(/^\.([^\s:,\[]+)/);
        if (!selectorMatch) return;
        const className = selectorMatch[1];
        const declarations = [];
        rule.walkDecls((decl) => {
          declarations.push(`${decl.prop}: ${decl.value}`);
        });
        if (declarations.length > 0) {
          utilityMap.set(className, declarations);
        }
      });
    } catch {
    }
  }
  return utilityMap;
}
async function generateConsolidatedCSS(mappings, buildDir, cssLayer) {
  const utilityMap = await extractUtilityCSS(buildDir);
  const cssRules = [];
  for (const mapping of mappings) {
    const declarations = [];
    for (const className of mapping.classes) {
      const extracted = utilityMap.get(className);
      if (extracted) {
        declarations.push(...extracted);
        continue;
      }
      const parsed = parseUtilityClass(className);
      if (parsed.length > 0) {
        declarations.push(...parsed);
      }
    }
    const uniqueDeclarations = [...new Set(declarations)];
    if (uniqueDeclarations.length > 0) {
      mapping.cssDeclarations = uniqueDeclarations.join("; ");
      cssRules.push(`.${mapping.consolidated} {
  ${uniqueDeclarations.join(";\n  ")};
}`);
    }
  }
  const cssContent = cssRules.join("\n\n");
  if (cssLayer) {
    return `/* Classpresso Consolidated Classes */
@layer ${cssLayer} {
${cssRules.map((rule) => "  " + rule.replace(/\n/g, "\n  ")).join("\n\n")}
}`;
  }
  return `/* Classpresso Consolidated Classes */
${cssContent}`;
}
async function injectConsolidatedCSS(buildDir, consolidatedCSS) {
  const cssPatterns = [
    // Next.js
    "static/**/*.css",
    "static/css/**/*.css",
    "standalone/.next/static/**/*.css",
    "standalone/.next/static/css/**/*.css",
    // Angular
    "browser/**/*.css",
    // Astro
    "_astro/**/*.css",
    "client/_astro/**/*.css",
    // Nuxt
    "public/_nuxt/**/*.css",
    // SvelteKit
    "_app/**/*.css",
    "client/**/*.css",
    // Vite/Vue/React/Qwik generic
    "assets/**/*.css",
    // Gatsby/Hugo/Eleventy/Docusaurus/Parcel
    "**/*.css"
  ];
  const cssFiles = await findFiles(buildDir, cssPatterns);
  if (cssFiles.length === 0) {
    throw new Error("No CSS files found in build output");
  }
  let targetFile = cssFiles[0];
  let maxSize = 0;
  for (const cssFile of cssFiles) {
    if (!isCSSFile(cssFile)) continue;
    try {
      const content = await readFileContent(cssFile);
      if (content.includes("Classpresso Consolidated")) {
        return `none (already injected in ${cssFile})`;
      }
      const size = Buffer.byteLength(content, "utf-8");
      if (size > maxSize) {
        maxSize = size;
        targetFile = cssFile;
      }
    } catch {
    }
  }
  const existingContent = await readFileContent(targetFile);
  const newContent = `${existingContent}

${consolidatedCSS}`;
  await writeFileContent(targetFile, newContent);
  return targetFile;
}

// src/core/transformer.ts
init_files();
function isSupersetOfDynamicBase(classes, dynamicBasePatterns) {
  const classSet = new Set(classes);
  for (const [, pattern] of dynamicBasePatterns) {
    const allBasePresent = pattern.baseClasses.every((cls) => classSet.has(cls));
    if (allBasePresent && classes.length > pattern.baseClasses.length) {
      return true;
    }
  }
  return false;
}
function matchesDynamicBase(classes, dynamicBasePatterns) {
  const classSet = new Set(classes);
  for (const [, pattern] of dynamicBasePatterns) {
    if (classes.length !== pattern.baseClasses.length) continue;
    const allMatch = pattern.baseClasses.every((cls) => classSet.has(cls));
    if (allMatch) return true;
  }
  return false;
}
function buildMatchPatterns(classString) {
  const escaped = escapeRegex(classString);
  return [
    // className="..." (JSX/HTML) - supports data attributes
    { regex: new RegExp(`(className\\s*=\\s*")${escaped}(")`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(className\\s*=\\s*')${escaped}(')`, "g"), supportsDataAttr: true },
    // className:"..." (React createElement) - no data attr support
    { regex: new RegExp(`(className\\s*:\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    { regex: new RegExp(`(className\\s*:\\s*')${escaped}(')`, "g"), supportsDataAttr: false },
    // Template literals: className:`...` (static content only)
    { regex: new RegExp(`(className\\s*:\\s*\`)${escaped}(\`)`, "g"), supportsDataAttr: false },
    // Template literals with dynamic suffix: className:`... ${...}`
    // Matches static prefix before ${...} expression
    { regex: new RegExp(`(className\\s*:\\s*\`)${escaped}( \\$\\{)`, "g"), supportsDataAttr: false },
    // "className","..." (minified) - no data attr support
    { regex: new RegExp(`("className"\\s*,\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    // class="..." (HTML) - supports data attributes
    { regex: new RegExp(`(\\bclass\\s*=\\s*")${escaped}(")`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(\\bclass\\s*=\\s*')${escaped}(')`, "g"), supportsDataAttr: true },
    // class:"..." (Vue/Quasar createElement) - no data attr support
    { regex: new RegExp(`(\\bclass\\s*:\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    { regex: new RegExp(`(\\bclass\\s*:\\s*')${escaped}(')`, "g"), supportsDataAttr: false },
    // HTML entity encoded: class=&quot;...&quot; - supports data attributes
    { regex: new RegExp(`(class=&quot;)${escaped}(&quot;)`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(class=&#34;)${escaped}(&#34;)`, "g"), supportsDataAttr: true },
    // RSC payload (escaped JSON in HTML): \"className\":\"...\"
    // This is used in Next.js self.__next_f.push() payloads
    { regex: new RegExp(`(\\\\"className\\\\"\\s*:\\s*\\\\")${escaped}(\\\\")`, "g"), supportsDataAttr: false }
  ];
}
function replacePattern(content, originalClasses, consolidated, excludedClasses, originalClassList, addDataAttributes) {
  let replacements = 0;
  let result = content;
  const replacement = excludedClasses.length > 0 ? `${consolidated} ${excludedClasses.join(" ")}` : consolidated;
  const dataAttr = addDataAttributes ? ` data-cp-original="${originalClassList.join(" ")}"` : "";
  const patterns = buildMatchPatterns(originalClasses);
  for (const { regex: pattern, supportsDataAttr } of patterns) {
    const beforeLength = result.length;
    const suffix = addDataAttributes && supportsDataAttr ? dataAttr : "";
    result = result.replace(pattern, `$1${replacement}$2${suffix}`);
    if (result.length !== beforeLength) {
      const matches = content.match(pattern);
      if (matches) replacements += matches.length;
    }
  }
  return { content: result, replacements };
}
function findClassVariations(content, mapping, config) {
  const variations = [];
  const targetClasses = new Set(mapping.classes);
  const classAttrPatterns = [
    /(?:className|class)\s*[=:]\s*["']([^"']+)["']/g,
    /class=&quot;([^&]+)&quot;/g,
    /class=&#34;([^&]+)&#34;/g,
    // Template literals - static only: className:`...`
    /(?:className|class)\s*:\s*`([^`$]+)`/g,
    // Template literals with dynamic suffix: className:`... ${...}`
    // Captures the static part before the ${
    /(?:className|class)\s*:\s*`([^`$]+) \$\{/g,
    // RSC payload (escaped JSON in HTML): \"className\":\"...\"
    // Used in Next.js self.__next_f.push() payloads
    /\\"className\\"\s*:\s*\\"([^"\\]+)\\"/g
  ];
  for (const classAttrRegex of classAttrPatterns) {
    classAttrRegex.lastIndex = 0;
    let match;
    while ((match = classAttrRegex.exec(content)) !== null) {
      const classString = match[1];
      const { classes } = normalizeClassString(classString, config.exclude);
      if (classes.length >= targetClasses.size) {
        const classSet = new Set(classes);
        let allPresent = true;
        for (const targetClass of targetClasses) {
          if (!classSet.has(targetClass)) {
            allPresent = false;
            break;
          }
        }
        if (allPresent && classes.length === targetClasses.size) {
          variations.push(classString);
        }
      }
    }
  }
  return [...new Set(variations)];
}
async function transformFile(filePath, mappings, config, dryRun, dynamicBasePatterns, mergeablePatterns) {
  try {
    let content = await readFileContent(filePath);
    let totalReplacements = 0;
    let skippedForHydration = 0;
    const originalContent = content;
    const isJS = isJSFile(filePath);
    for (const mapping of mappings) {
      if (dynamicBasePatterns.size > 0) {
        const isSuperset = isSupersetOfDynamicBase(mapping.classes, dynamicBasePatterns);
        const isExact = matchesDynamicBase(mapping.classes, dynamicBasePatterns);
        if (isSuperset || isExact) {
          skippedForHydration++;
          continue;
        }
      }
      if (isJS && mergeablePatterns.has(mapping.original)) {
        skippedForHydration++;
        continue;
      }
      const variations = findClassVariations(content, mapping, config);
      for (const variation of variations) {
        const { content: newContent, replacements } = replacePattern(
          content,
          variation,
          mapping.consolidated,
          mapping.excludedClasses,
          mapping.classes,
          config.dataAttributes
        );
        content = newContent;
        totalReplacements += replacements;
      }
    }
    const modified = content !== originalContent;
    if (modified && !dryRun) {
      if (config.backup) {
        await backupFile(filePath);
      }
      await writeFileContent(filePath, content);
    }
    return { modified, replacements: totalReplacements, skippedForHydration };
  } catch (err) {
    return {
      modified: false,
      replacements: 0,
      skippedForHydration: 0,
      error: `Error transforming ${filePath}: ${err}`
    };
  }
}
async function transformBuildOutput(mappings, config, dryRun = false, dynamicBasePatterns = /* @__PURE__ */ new Map(), mergeablePatterns = /* @__PURE__ */ new Set()) {
  if (config.verbose) {
    if (dynamicBasePatterns.size > 0) {
      console.log(`  \u2139 Found ${dynamicBasePatterns.size} dynamic base patterns (template literals)`);
    }
    if (mergeablePatterns.size > 0) {
      console.log(`  \u2139 Found ${mergeablePatterns.size} mergeable patterns (className props)`);
    }
  }
  const patterns = config.include.length > 0 ? config.include : DEFAULT_PATTERNS;
  const allFiles = await findFiles(config.buildDir, patterns);
  const files = filterExcludedFiles(allFiles, config.exclude.files || []);
  if (config.verbose && allFiles.length !== files.length) {
    console.log(`  \u2139 Excluded ${allFiles.length - files.length} files from transformation`);
  }
  let filesModified = 0;
  let totalBytesChanged = 0;
  let totalSkippedForHydration = 0;
  const errors = [];
  for (const filePath of files) {
    if (!isJSFile(filePath) && !isHTMLFile(filePath) && !isRSCFile(filePath)) {
      continue;
    }
    const result = await transformFile(filePath, mappings, config, dryRun, dynamicBasePatterns, mergeablePatterns);
    if (result.error) {
      errors.push(result.error);
    }
    if (result.modified) {
      filesModified++;
      totalBytesChanged += result.replacements * 10;
    }
    totalSkippedForHydration += result.skippedForHydration;
  }
  if (config.verbose && totalSkippedForHydration > 0) {
    console.log(`Skipped ${totalSkippedForHydration} pattern transformations to prevent hydration mismatches`);
  }
  return {
    filesModified,
    bytesChanged: totalBytesChanged,
    errors
  };
}

// src/config.ts
init_files();
var DEFAULT_DYNAMIC_PREFIXES = [
  "lucide",
  // lucide-react icons
  "heroicon",
  // heroicons
  "icon-",
  // generic icon patterns
  "fa-",
  // Font Awesome
  "fas",
  // Font Awesome solid
  "far",
  // Font Awesome regular
  "fab",
  // Font Awesome brands
  "material-icons",
  // Material Design Icons
  "mdi-",
  // Material Design Icons (alternative)
  "ri-",
  // Remix Icons
  "bi-",
  // Bootstrap Icons
  "tabler-",
  // Tabler Icons
  "phosphor-"
  // Phosphor Icons
];
var DEFAULT_CONFIG = {
  buildDir: ".next",
  minOccurrences: 2,
  minClasses: 2,
  minBytesSaved: 10,
  hashPrefix: "cp-",
  hashLength: 5,
  exclude: {
    prefixes: ["js-", "data-", "hook-", "track-"],
    suffixes: ["-handler", "-trigger"],
    classes: [],
    patterns: [/^qa-/, /^test-/, /^e2e-/],
    files: []
  },
  include: [],
  cssLayer: false,
  dataAttributes: false,
  manifest: true,
  backup: false,
  verbose: false,
  forceAll: false,
  excludeDynamicPatterns: true,
  dynamicPrefixes: DEFAULT_DYNAMIC_PREFIXES,
  skipPatternsWithExcludedClasses: true,
  ssr: false,
  debug: false,
  sendErrorReports: false,
  errorReportUrl: void 0,
  purgeUnusedCSS: false,
  purgeSafelist: [
    /^js-/,
    // JavaScript hooks
    /^data-/,
    // Data attributes used as classes
    /^is-/,
    // State classes
    /^has-/,
    // State classes
    /^active$/,
    /^disabled$/,
    /^hidden$/,
    /^visible$/
  ]
};
async function loadConfig(buildDir) {
  const config = { ...DEFAULT_CONFIG };
  if (buildDir) {
    config.buildDir = buildDir;
  }
  const configPaths = [
    "classpresso.config.js",
    "classpresso.config.mjs",
    "classpresso.config.json"
  ];
  for (const configPath of configPaths) {
    try {
      const fullPath = path2.resolve(process.cwd(), configPath);
      if (configPath.endsWith(".json")) {
        const content = await readFileContent(fullPath);
        const userConfig = JSON.parse(content);
        return mergeConfig(config, userConfig);
      } else {
        const userConfig = await import(fullPath);
        return mergeConfig(config, userConfig.default || userConfig);
      }
    } catch {
    }
  }
  return config;
}
function mergeConfig(defaults, userConfig) {
  return {
    ...defaults,
    ...userConfig,
    exclude: {
      ...defaults.exclude,
      ...userConfig.exclude,
      // Merge arrays instead of replacing
      prefixes: [
        ...defaults.exclude.prefixes || [],
        ...userConfig.exclude?.prefixes || []
      ],
      suffixes: [
        ...defaults.exclude.suffixes || [],
        ...userConfig.exclude?.suffixes || []
      ],
      classes: [
        ...defaults.exclude.classes || [],
        ...userConfig.exclude?.classes || []
      ],
      patterns: [
        ...defaults.exclude.patterns || [],
        ...userConfig.exclude?.patterns || []
      ],
      files: [
        ...defaults.exclude.files || [],
        ...userConfig.exclude?.files || []
      ]
    },
    // Merge dynamicPrefixes arrays
    dynamicPrefixes: [
      ...defaults.dynamicPrefixes || [],
      ...userConfig.dynamicPrefixes || []
    ],
    // Merge purgeSafelist arrays
    purgeSafelist: [
      ...defaults.purgeSafelist || [],
      ...userConfig.purgeSafelist || []
    ]
  };
}
function validateConfig(config) {
  const errors = [];
  if (config.minOccurrences < 1) {
    errors.push("minOccurrences must be at least 1");
  }
  if (config.minClasses < 1) {
    errors.push("minClasses must be at least 1");
  }
  if (config.hashLength < 3) {
    errors.push("hashLength must be at least 3");
  }
  if (config.hashLength > 32) {
    errors.push("hashLength must be at most 32");
  }
  if (!config.hashPrefix || config.hashPrefix.length === 0) {
    errors.push("hashPrefix must not be empty");
  }
  if (config.sendErrorReports && !config.errorReportUrl) {
    errors.push("errorReportUrl is required when sendErrorReports is true");
  }
  if (config.errorReportUrl && !isValidHttpsUrl(config.errorReportUrl)) {
    errors.push("errorReportUrl must be a valid HTTPS URL");
  }
  return errors;
}
function isValidHttpsUrl(url) {
  try {
    const parsed = new URL(url);
    return parsed.protocol === "https:";
  } catch {
    return false;
  }
}

// src/index.ts
init_files();

export { ALL_CLASS_PATTERNS, CLASS_PATTERNS, DEFAULT_CONFIG, DEFAULT_DYNAMIC_PREFIXES, DEFAULT_PATTERNS, DYNAMIC_BASE_PATTERNS, backupFile, buildReplacementMap, calculateMetrics, containsDynamicPrefix, createClassMappings, detectConsolidatablePatterns, detectMergeablePatterns, escapeRegex, estimateCSSOverhead, extractDynamicBaseStrings, findFiles, formatBytes, formatPercentage, formatTime, generateConsolidatedCSS, generateHashName, getFileSize, getPatternSummary, injectConsolidatedCSS, isDynamicClassString, isProperSubset, loadConfig, loadMappingManifest, normalizeClassString, parseUtilityClass, readFileContent, resolveCollisions, saveMappingManifest, scanBuildOutput, shouldExcludeClass, transformBuildOutput, validateConfig, writeFileContent };
