// classpresso/dist/scanner.cjs
'use strict';

var promises = require('fs/promises');
var glob = require('glob');
var path = require('path');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var path__default = /*#__PURE__*/_interopDefault(path);

// src/utils/files.ts
var DEFAULT_PATTERNS = [
  // === Next.js (.next directory) ===
  // Client-side chunks
  "static/chunks/**/*.js",
  "static/css/**/*.css",
  // Server components (Next.js App Router)
  "server/app/**/*.js",
  "server/app/**/*.html",
  "server/app/**/*.rsc",
  // Server chunks (contains Client Component code used for hydration)
  "server/chunks/**/*.js",
  // Standalone build (Next.js output: 'standalone')
  "standalone/.next/static/chunks/**/*.js",
  "standalone/.next/static/css/**/*.css",
  "standalone/.next/server/app/**/*.js",
  "standalone/.next/server/app/**/*.html",
  "standalone/.next/server/app/**/*.rsc",
  "standalone/.next/server/chunks/**/*.js",
  // === Nuxt (.output directory) ===
  "public/**/*.html",
  "public/**/*.js",
  "public/**/*.css",
  "public/_nuxt/**/*.js",
  "public/_nuxt/**/*.css",
  "server/**/*.mjs",
  "server/chunks/**/*.mjs",
  // === SvelteKit (build directory) ===
  // Static adapter output
  "**/*.html",
  "_app/**/*.js",
  "_app/**/*.css",
  // Node adapter output
  "client/**/*.js",
  "client/**/*.css",
  "server/**/*.js",
  // === Astro (dist directory) ===
  "_astro/**/*.js",
  "_astro/**/*.css",
  // SSR/Hybrid build - Client assets
  "client/_astro/**/*.js",
  "client/_astro/**/*.css",
  // === Remix (build directory) ===
  "client/**/*.js",
  "client/**/*.css",
  "server/**/*.js",
  // === Solid Start (.output or dist directory) ===
  "public/**/*.html",
  "public/**/*.js",
  "public/**/*.css",
  "_build/**/*.js",
  "_build/**/*.css",
  // === Qwik (dist directory) ===
  "build/**/*.js",
  "build/**/*.css",
  "q-*.js",
  // === Angular (dist/[project] directory) ===
  // Angular 17+ uses browser subdirectory
  "browser/**/*.js",
  "browser/**/*.css",
  "browser/**/*.html",
  // Legacy Angular output
  "*.js",
  "*.css",
  "*.html",
  // === Gatsby (public directory) ===
  "page-data/**/*.json",
  "*.js",
  "*.css",
  "*.html",
  // === Vite/Vue/React generic (dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "index.html",
  // === Ember (dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "*.html",
  // === VitePress (.vitepress/dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "**/*.html",
  // === Docusaurus (build directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "**/*.html",
  // === Eleventy/11ty (_site directory) ===
  "**/*.html",
  "**/*.js",
  "**/*.css",
  // === Hugo (public directory) ===
  "**/*.html",
  "**/*.js",
  "**/*.css",
  // === Parcel (dist directory) ===
  "*.html",
  "*.js",
  "*.css",
  // === Preact (build directory) ===
  "bundle.js",
  "bundle.*.js",
  "*.css",
  "*.html",
  // === Gridsome (dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "**/*.html",
  // === RedwoodJS (web/dist directory) ===
  "static/**/*.js",
  "static/**/*.css",
  "*.html",
  // === Webpack generic (dist directory) ===
  "*.js",
  "*.css",
  "*.html",
  "js/**/*.js",
  "css/**/*.css"
];
async function findFiles(buildDir, patterns) {
  const globPatterns = [...new Set(patterns.map((p) => path__default.default.join(buildDir, p)))];
  const allFiles = await glob.glob(globPatterns, {
    nodir: true,
    windowsPathsNoEscape: true
  });
  return [...new Set(allFiles)];
}
async function readFileContent(filePath) {
  return promises.readFile(filePath, "utf-8");
}
async function getFileSize(filePath) {
  const stats = await promises.stat(filePath);
  return stats.size;
}
function isJSFile(filePath) {
  return /\.(js|mjs|cjs)$/.test(filePath);
}
function isHTMLFile(filePath) {
  return /\.(html|htm)$/.test(filePath);
}
function isRSCFile(filePath) {
  return /\.rsc$/.test(filePath);
}
function shouldExcludeFile(filePath, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return false;
  }
  const normalizedPath = filePath.replace(/\\/g, "/");
  for (const pattern of excludePatterns) {
    const regexPattern = pattern.replace(/\\/g, "/").replace(/\*\*/g, "{{DOUBLESTAR}}").replace(/\*/g, "[^/]*").replace(/\?/g, ".").replace(/{{DOUBLESTAR}}/g, ".*");
    const regex = new RegExp(regexPattern);
    if (regex.test(normalizedPath)) {
      return true;
    }
  }
  return false;
}
function filterExcludedFiles(files, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return files;
  }
  return files.filter((filePath) => !shouldExcludeFile(filePath, excludePatterns));
}

// src/utils/regex.ts
var CLASS_PATTERNS = {
  // JSX: className="..."
  jsxDouble: /className\s*=\s*"([^"]+)"/g,
  // JSX: className='...'
  jsxSingle: /className\s*=\s*'([^']+)'/g,
  // React createElement: className:"..."
  createElementDouble: /className\s*:\s*"([^"]+)"/g,
  // React createElement: className:'...'
  createElementSingle: /className\s*:\s*'([^']+)'/g,
  // Minified: "className","..."
  minifiedComma: /"className"\s*,\s*"([^"]+)"/g,
  // HTML: class="..."
  htmlDouble: /\bclass\s*=\s*"([^"]+)"/g,
  // HTML: class='...'
  htmlSingle: /\bclass\s*=\s*'([^']+)'/g,
  // HTML entity encoded: class=&quot;...&quot;
  htmlEntityDouble: /\bclass=&quot;([^&]+)&quot;/g,
  // HTML entity encoded: class=&#34;...&#34; (numeric entity)
  htmlNumericDouble: /\bclass=&#34;([^&]+)&#34;/g,
  // RSC payload: \"className\":\"...\" (escaped JSON in Next.js RSC payloads)
  rscPayload: /\\"className\\"\s*:\s*\\"([^"\\]+)\\"/g,
  // Vue/Quasar createElement: class:"..." (Vue uses 'class' not 'className')
  vueCreateElementDouble: /\bclass\s*:\s*"([^"]+)"/g,
  // Vue/Quasar createElement: class:'...'
  vueCreateElementSingle: /\bclass\s*:\s*'([^']+)'/g
};
var ALL_CLASS_PATTERNS = [
  CLASS_PATTERNS.jsxDouble,
  CLASS_PATTERNS.jsxSingle,
  CLASS_PATTERNS.createElementDouble,
  CLASS_PATTERNS.createElementSingle,
  CLASS_PATTERNS.minifiedComma,
  CLASS_PATTERNS.htmlDouble,
  CLASS_PATTERNS.htmlSingle,
  CLASS_PATTERNS.htmlEntityDouble,
  CLASS_PATTERNS.htmlNumericDouble,
  CLASS_PATTERNS.rscPayload,
  CLASS_PATTERNS.vueCreateElementDouble,
  CLASS_PATTERNS.vueCreateElementSingle
];
function isDynamicClassString(classString) {
  if (classString.includes("${")) return true;
  if (/\s*\?\s*/.test(classString) && classString.includes(":")) return true;
  if (/\w+\s*\(/.test(classString)) return true;
  if (/^[a-z][a-zA-Z0-9]*$/.test(classString) && !classString.includes("-")) return true;
  return false;
}
var DYNAMIC_BASE_PATTERNS = [
  // className:`base classes ${...}`
  /className\s*:\s*`([^`$]+)\s*\$\{/g,
  // className=`base classes ${...}` (JSX)
  /className\s*=\s*`([^`$]+)\s*\$\{/g,
  // class:`base classes ${...}` (rare but possible)
  /\bclass\s*:\s*`([^`$]+)\s*\$\{/g
];
var CONCAT_BASE_PATTERNS = [
  // className:"".concat("base classes",...) - the base before concat
  /className\s*:\s*""\s*\.concat\s*\(\s*"([^"]+)"/g,
  // className:''.concat('base classes',...)
  /className\s*:\s*''\s*\.concat\s*\(\s*'([^']+)'/g
];
function extractConcatBaseStrings(content) {
  const results = [];
  for (const pattern of CONCAT_BASE_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const baseString = match[1]?.trim();
      if (baseString && baseString.length > 0) {
        results.push(baseString);
      }
    }
  }
  return results;
}
function extractDynamicBaseStrings(content) {
  const results = [];
  for (const pattern of DYNAMIC_BASE_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const baseString = match[1]?.trim();
      if (baseString && baseString.length > 0) {
        results.push(baseString);
      }
    }
  }
  return results;
}

// src/core/scanner.ts
function getSourceFileType(filePath) {
  if (isHTMLFile(filePath)) return "html";
  if (isRSCFile(filePath)) return "rsc";
  return "js";
}
function isProperSubset(classesA, classesB) {
  if (classesA.length >= classesB.length) return false;
  const setB = new Set(classesB);
  return classesA.every((cls) => setB.has(cls));
}
function detectMergeablePatterns(occurrences) {
  const mergeablePatterns = /* @__PURE__ */ new Set();
  const jsPatterns = [];
  const htmlPatterns = [];
  for (const [, occurrence] of occurrences) {
    if (occurrence.sourceTypes.has("js")) {
      jsPatterns.push(occurrence);
    }
    if (occurrence.sourceTypes.has("html") || occurrence.sourceTypes.has("rsc")) {
      htmlPatterns.push(occurrence);
    }
  }
  for (const jsPattern of jsPatterns) {
    for (const htmlPattern of htmlPatterns) {
      if (isProperSubset(jsPattern.classes, htmlPattern.classes)) {
        mergeablePatterns.add(jsPattern.normalizedKey);
        break;
      }
    }
  }
  return mergeablePatterns;
}
function containsDynamicPrefix(classes, dynamicPrefixes) {
  for (const cls of classes) {
    for (const prefix of dynamicPrefixes) {
      if (prefix.endsWith("-")) {
        if (cls.startsWith(prefix)) {
          return true;
        }
      } else {
        if (cls === prefix || cls.startsWith(prefix + "-")) {
          return true;
        }
      }
    }
  }
  return false;
}
function isHydrationSafe(occurrence) {
  const hasServerContext = occurrence.sourceTypes.has("html") || occurrence.sourceTypes.has("rsc");
  const hasClientContext = occurrence.sourceTypes.has("js");
  return hasServerContext && hasClientContext;
}
function shouldExcludeClass(className, exclude) {
  if (exclude.prefixes) {
    for (const prefix of exclude.prefixes) {
      if (className.startsWith(prefix)) return true;
    }
  }
  if (exclude.suffixes) {
    for (const suffix of exclude.suffixes) {
      if (className.endsWith(suffix)) return true;
    }
  }
  if (exclude.classes) {
    if (exclude.classes.includes(className)) return true;
  }
  if (exclude.patterns) {
    for (const pattern of exclude.patterns) {
      if (pattern.test(className)) return true;
    }
  }
  return false;
}
function normalizeClassString(classString, exclude) {
  const allClasses = classString.split(/\s+/).filter(Boolean);
  const includedClasses = [];
  const excludedClasses = [];
  for (const cls of allClasses) {
    if (shouldExcludeClass(cls, exclude)) {
      excludedClasses.push(cls);
    } else {
      includedClasses.push(cls);
    }
  }
  const sorted = [...includedClasses].sort();
  return {
    normalized: sorted.join(" "),
    classes: includedClasses,
    excludedClasses
  };
}
function extractClassStrings(content, filePath) {
  const results = [];
  const seen = /* @__PURE__ */ new Set();
  for (const pattern of ALL_CLASS_PATTERNS) {
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const classString = match[1];
      if (!classString || !classString.trim()) continue;
      if (isDynamicClassString(classString)) continue;
      const key = `${filePath}:${match.index}:${classString}`;
      if (seen.has(key)) continue;
      seen.add(key);
      const beforeMatch = content.substring(0, match.index);
      const line = (beforeMatch.match(/\n/g) || []).length + 1;
      results.push({
        classString: classString.trim(),
        location: {
          filePath,
          line
        }
      });
    }
  }
  return results;
}
async function scanBuildOutput(config) {
  const occurrences = /* @__PURE__ */ new Map();
  const dynamicBasePatterns = /* @__PURE__ */ new Map();
  const fileStats = [];
  const errors = [];
  const patterns = config.include.length > 0 ? config.include : DEFAULT_PATTERNS;
  const allFiles = await findFiles(config.buildDir, patterns);
  const files = filterExcludedFiles(allFiles, config.exclude.files || []);
  if (config.verbose) {
    console.log(`Found ${allFiles.length} files, scanning ${files.length} (${allFiles.length - files.length} excluded)`);
  }
  for (const filePath of files) {
    try {
      if (!isJSFile(filePath) && !isHTMLFile(filePath) && !isRSCFile(filePath)) {
        continue;
      }
      const content = await readFileContent(filePath);
      const size = await getFileSize(filePath);
      fileStats.push({
        path: filePath,
        originalSize: size,
        modified: false
      });
      if (isJSFile(filePath)) {
        const dynamicBases = extractDynamicBaseStrings(content);
        const concatBases = extractConcatBaseStrings(content);
        const allDynamicBases = [...dynamicBases, ...concatBases];
        for (const baseString of allDynamicBases) {
          const { normalized, classes } = normalizeClassString(baseString, config.exclude);
          if (classes.length < config.minClasses) continue;
          const location = { filePath };
          if (dynamicBasePatterns.has(normalized)) {
            dynamicBasePatterns.get(normalized).locations.push(location);
          } else {
            dynamicBasePatterns.set(normalized, {
              baseClasses: classes,
              normalizedKey: normalized,
              locations: [location]
            });
          }
        }
      }
      const classStrings = extractClassStrings(content, filePath);
      const sourceType = getSourceFileType(filePath);
      for (const { classString, location } of classStrings) {
        const { normalized, classes, excludedClasses } = normalizeClassString(
          classString,
          config.exclude
        );
        if (classes.length === 0) continue;
        if (classes.length < config.minClasses) continue;
        if (config.excludeDynamicPatterns && containsDynamicPrefix(classes, config.dynamicPrefixes)) {
          continue;
        }
        if (occurrences.has(normalized)) {
          const existing = occurrences.get(normalized);
          existing.count++;
          existing.locations.push(location);
          existing.sourceTypes.add(sourceType);
        } else {
          occurrences.set(normalized, {
            classString,
            normalizedKey: normalized,
            count: 1,
            locations: [location],
            classes,
            excludedClasses,
            sourceTypes: /* @__PURE__ */ new Set([sourceType])
          });
        }
      }
    } catch (err) {
      errors.push(`Error scanning ${filePath}: ${err}`);
    }
  }
  const mergeablePatterns = detectMergeablePatterns(occurrences);
  if (config.verbose && mergeablePatterns.size > 0) {
    console.log(`Found ${mergeablePatterns.size} patterns that appear to be merged className props`);
  }
  return {
    occurrences,
    files: fileStats,
    errors,
    dynamicBasePatterns,
    mergeablePatterns
  };
}

exports.containsDynamicPrefix = containsDynamicPrefix;
exports.detectMergeablePatterns = detectMergeablePatterns;
exports.extractClassStrings = extractClassStrings;
exports.isHydrationSafe = isHydrationSafe;
exports.isProperSubset = isProperSubset;
exports.normalizeClassString = normalizeClassString;
exports.scanBuildOutput = scanBuildOutput;
exports.shouldExcludeClass = shouldExcludeClass;
