// classpresso/dist/transformer.js
import { readFile, copyFile, writeFile } from 'fs/promises';
import { glob } from 'glob';
import path from 'path';

// src/utils/files.ts
var DEFAULT_PATTERNS = [
  // === Next.js (.next directory) ===
  // Client-side chunks
  "static/chunks/**/*.js",
  "static/css/**/*.css",
  // Server components (Next.js App Router)
  "server/app/**/*.js",
  "server/app/**/*.html",
  "server/app/**/*.rsc",
  // Server chunks (contains Client Component code used for hydration)
  "server/chunks/**/*.js",
  // Standalone build (Next.js output: 'standalone')
  "standalone/.next/static/chunks/**/*.js",
  "standalone/.next/static/css/**/*.css",
  "standalone/.next/server/app/**/*.js",
  "standalone/.next/server/app/**/*.html",
  "standalone/.next/server/app/**/*.rsc",
  "standalone/.next/server/chunks/**/*.js",
  // === Nuxt (.output directory) ===
  "public/**/*.html",
  "public/**/*.js",
  "public/**/*.css",
  "public/_nuxt/**/*.js",
  "public/_nuxt/**/*.css",
  "server/**/*.mjs",
  "server/chunks/**/*.mjs",
  // === SvelteKit (build directory) ===
  // Static adapter output
  "**/*.html",
  "_app/**/*.js",
  "_app/**/*.css",
  // Node adapter output
  "client/**/*.js",
  "client/**/*.css",
  "server/**/*.js",
  // === Astro (dist directory) ===
  "_astro/**/*.js",
  "_astro/**/*.css",
  // SSR/Hybrid build - Client assets
  "client/_astro/**/*.js",
  "client/_astro/**/*.css",
  // === Remix (build directory) ===
  "client/**/*.js",
  "client/**/*.css",
  "server/**/*.js",
  // === Solid Start (.output or dist directory) ===
  "public/**/*.html",
  "public/**/*.js",
  "public/**/*.css",
  "_build/**/*.js",
  "_build/**/*.css",
  // === Qwik (dist directory) ===
  "build/**/*.js",
  "build/**/*.css",
  "q-*.js",
  // === Angular (dist/[project] directory) ===
  // Angular 17+ uses browser subdirectory
  "browser/**/*.js",
  "browser/**/*.css",
  "browser/**/*.html",
  // Legacy Angular output
  "*.js",
  "*.css",
  "*.html",
  // === Gatsby (public directory) ===
  "page-data/**/*.json",
  "*.js",
  "*.css",
  "*.html",
  // === Vite/Vue/React generic (dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "index.html",
  // === Ember (dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "*.html",
  // === VitePress (.vitepress/dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "**/*.html",
  // === Docusaurus (build directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "**/*.html",
  // === Eleventy/11ty (_site directory) ===
  "**/*.html",
  "**/*.js",
  "**/*.css",
  // === Hugo (public directory) ===
  "**/*.html",
  "**/*.js",
  "**/*.css",
  // === Parcel (dist directory) ===
  "*.html",
  "*.js",
  "*.css",
  // === Preact (build directory) ===
  "bundle.js",
  "bundle.*.js",
  "*.css",
  "*.html",
  // === Gridsome (dist directory) ===
  "assets/**/*.js",
  "assets/**/*.css",
  "**/*.html",
  // === RedwoodJS (web/dist directory) ===
  "static/**/*.js",
  "static/**/*.css",
  "*.html",
  // === Webpack generic (dist directory) ===
  "*.js",
  "*.css",
  "*.html",
  "js/**/*.js",
  "css/**/*.css"
];
async function findFiles(buildDir, patterns) {
  const globPatterns = [...new Set(patterns.map((p) => path.join(buildDir, p)))];
  const allFiles = await glob(globPatterns, {
    nodir: true,
    windowsPathsNoEscape: true
  });
  return [...new Set(allFiles)];
}
async function readFileContent(filePath) {
  return readFile(filePath, "utf-8");
}
async function writeFileContent(filePath, content) {
  await writeFile(filePath, content, "utf-8");
}
async function backupFile(filePath) {
  const backupPath = `${filePath}.classpresso-backup`;
  await copyFile(filePath, backupPath);
  return backupPath;
}
function isJSFile(filePath) {
  return /\.(js|mjs|cjs)$/.test(filePath);
}
function isHTMLFile(filePath) {
  return /\.(html|htm)$/.test(filePath);
}
function isRSCFile(filePath) {
  return /\.rsc$/.test(filePath);
}
function shouldExcludeFile(filePath, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return false;
  }
  const normalizedPath = filePath.replace(/\\/g, "/");
  for (const pattern of excludePatterns) {
    const regexPattern = pattern.replace(/\\/g, "/").replace(/\*\*/g, "{{DOUBLESTAR}}").replace(/\*/g, "[^/]*").replace(/\?/g, ".").replace(/{{DOUBLESTAR}}/g, ".*");
    const regex = new RegExp(regexPattern);
    if (regex.test(normalizedPath)) {
      return true;
    }
  }
  return false;
}
function filterExcludedFiles(files, excludePatterns) {
  if (!excludePatterns || excludePatterns.length === 0) {
    return files;
  }
  return files.filter((filePath) => !shouldExcludeFile(filePath, excludePatterns));
}

// src/utils/regex.ts
function escapeRegex(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

// src/core/scanner.ts
function shouldExcludeClass(className, exclude) {
  if (exclude.prefixes) {
    for (const prefix of exclude.prefixes) {
      if (className.startsWith(prefix)) return true;
    }
  }
  if (exclude.suffixes) {
    for (const suffix of exclude.suffixes) {
      if (className.endsWith(suffix)) return true;
    }
  }
  if (exclude.classes) {
    if (exclude.classes.includes(className)) return true;
  }
  if (exclude.patterns) {
    for (const pattern of exclude.patterns) {
      if (pattern.test(className)) return true;
    }
  }
  return false;
}
function normalizeClassString(classString, exclude) {
  const allClasses = classString.split(/\s+/).filter(Boolean);
  const includedClasses = [];
  const excludedClasses = [];
  for (const cls of allClasses) {
    if (shouldExcludeClass(cls, exclude)) {
      excludedClasses.push(cls);
    } else {
      includedClasses.push(cls);
    }
  }
  const sorted = [...includedClasses].sort();
  return {
    normalized: sorted.join(" "),
    classes: includedClasses,
    excludedClasses
  };
}

// src/core/transformer.ts
function isSupersetOfDynamicBase(classes, dynamicBasePatterns) {
  const classSet = new Set(classes);
  for (const [, pattern] of dynamicBasePatterns) {
    const allBasePresent = pattern.baseClasses.every((cls) => classSet.has(cls));
    if (allBasePresent && classes.length > pattern.baseClasses.length) {
      return true;
    }
  }
  return false;
}
function matchesDynamicBase(classes, dynamicBasePatterns) {
  const classSet = new Set(classes);
  for (const [, pattern] of dynamicBasePatterns) {
    if (classes.length !== pattern.baseClasses.length) continue;
    const allMatch = pattern.baseClasses.every((cls) => classSet.has(cls));
    if (allMatch) return true;
  }
  return false;
}
function buildMatchPatterns(classString) {
  const escaped = escapeRegex(classString);
  return [
    // className="..." (JSX/HTML) - supports data attributes
    { regex: new RegExp(`(className\\s*=\\s*")${escaped}(")`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(className\\s*=\\s*')${escaped}(')`, "g"), supportsDataAttr: true },
    // className:"..." (React createElement) - no data attr support
    { regex: new RegExp(`(className\\s*:\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    { regex: new RegExp(`(className\\s*:\\s*')${escaped}(')`, "g"), supportsDataAttr: false },
    // Template literals: className:`...` (static content only)
    { regex: new RegExp(`(className\\s*:\\s*\`)${escaped}(\`)`, "g"), supportsDataAttr: false },
    // Template literals with dynamic suffix: className:`... ${...}`
    // Matches static prefix before ${...} expression
    { regex: new RegExp(`(className\\s*:\\s*\`)${escaped}( \\$\\{)`, "g"), supportsDataAttr: false },
    // "className","..." (minified) - no data attr support
    { regex: new RegExp(`("className"\\s*,\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    // class="..." (HTML) - supports data attributes
    { regex: new RegExp(`(\\bclass\\s*=\\s*")${escaped}(")`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(\\bclass\\s*=\\s*')${escaped}(')`, "g"), supportsDataAttr: true },
    // class:"..." (Vue/Quasar createElement) - no data attr support
    { regex: new RegExp(`(\\bclass\\s*:\\s*")${escaped}(")`, "g"), supportsDataAttr: false },
    { regex: new RegExp(`(\\bclass\\s*:\\s*')${escaped}(')`, "g"), supportsDataAttr: false },
    // HTML entity encoded: class=&quot;...&quot; - supports data attributes
    { regex: new RegExp(`(class=&quot;)${escaped}(&quot;)`, "g"), supportsDataAttr: true },
    { regex: new RegExp(`(class=&#34;)${escaped}(&#34;)`, "g"), supportsDataAttr: true },
    // RSC payload (escaped JSON in HTML): \"className\":\"...\"
    // This is used in Next.js self.__next_f.push() payloads
    { regex: new RegExp(`(\\\\"className\\\\"\\s*:\\s*\\\\")${escaped}(\\\\")`, "g"), supportsDataAttr: false }
  ];
}
function replacePattern(content, originalClasses, consolidated, excludedClasses, originalClassList, addDataAttributes) {
  let replacements = 0;
  let result = content;
  const replacement = excludedClasses.length > 0 ? `${consolidated} ${excludedClasses.join(" ")}` : consolidated;
  const dataAttr = addDataAttributes ? ` data-cp-original="${originalClassList.join(" ")}"` : "";
  const patterns = buildMatchPatterns(originalClasses);
  for (const { regex: pattern, supportsDataAttr } of patterns) {
    const beforeLength = result.length;
    const suffix = addDataAttributes && supportsDataAttr ? dataAttr : "";
    result = result.replace(pattern, `$1${replacement}$2${suffix}`);
    if (result.length !== beforeLength) {
      const matches = content.match(pattern);
      if (matches) replacements += matches.length;
    }
  }
  return { content: result, replacements };
}
function findClassVariations(content, mapping, config) {
  const variations = [];
  const targetClasses = new Set(mapping.classes);
  const classAttrPatterns = [
    /(?:className|class)\s*[=:]\s*["']([^"']+)["']/g,
    /class=&quot;([^&]+)&quot;/g,
    /class=&#34;([^&]+)&#34;/g,
    // Template literals - static only: className:`...`
    /(?:className|class)\s*:\s*`([^`$]+)`/g,
    // Template literals with dynamic suffix: className:`... ${...}`
    // Captures the static part before the ${
    /(?:className|class)\s*:\s*`([^`$]+) \$\{/g,
    // RSC payload (escaped JSON in HTML): \"className\":\"...\"
    // Used in Next.js self.__next_f.push() payloads
    /\\"className\\"\s*:\s*\\"([^"\\]+)\\"/g
  ];
  for (const classAttrRegex of classAttrPatterns) {
    classAttrRegex.lastIndex = 0;
    let match;
    while ((match = classAttrRegex.exec(content)) !== null) {
      const classString = match[1];
      const { classes } = normalizeClassString(classString, config.exclude);
      if (classes.length >= targetClasses.size) {
        const classSet = new Set(classes);
        let allPresent = true;
        for (const targetClass of targetClasses) {
          if (!classSet.has(targetClass)) {
            allPresent = false;
            break;
          }
        }
        if (allPresent && classes.length === targetClasses.size) {
          variations.push(classString);
        }
      }
    }
  }
  return [...new Set(variations)];
}
async function transformFile(filePath, mappings, config, dryRun, dynamicBasePatterns, mergeablePatterns) {
  try {
    let content = await readFileContent(filePath);
    let totalReplacements = 0;
    let skippedForHydration = 0;
    const originalContent = content;
    const isJS = isJSFile(filePath);
    for (const mapping of mappings) {
      if (dynamicBasePatterns.size > 0) {
        const isSuperset = isSupersetOfDynamicBase(mapping.classes, dynamicBasePatterns);
        const isExact = matchesDynamicBase(mapping.classes, dynamicBasePatterns);
        if (isSuperset || isExact) {
          skippedForHydration++;
          continue;
        }
      }
      if (isJS && mergeablePatterns.has(mapping.original)) {
        skippedForHydration++;
        continue;
      }
      const variations = findClassVariations(content, mapping, config);
      for (const variation of variations) {
        const { content: newContent, replacements } = replacePattern(
          content,
          variation,
          mapping.consolidated,
          mapping.excludedClasses,
          mapping.classes,
          config.dataAttributes
        );
        content = newContent;
        totalReplacements += replacements;
      }
    }
    const modified = content !== originalContent;
    if (modified && !dryRun) {
      if (config.backup) {
        await backupFile(filePath);
      }
      await writeFileContent(filePath, content);
    }
    return { modified, replacements: totalReplacements, skippedForHydration };
  } catch (err) {
    return {
      modified: false,
      replacements: 0,
      skippedForHydration: 0,
      error: `Error transforming ${filePath}: ${err}`
    };
  }
}
async function transformBuildOutput(mappings, config, dryRun = false, dynamicBasePatterns = /* @__PURE__ */ new Map(), mergeablePatterns = /* @__PURE__ */ new Set()) {
  if (config.verbose) {
    if (dynamicBasePatterns.size > 0) {
      console.log(`  \u2139 Found ${dynamicBasePatterns.size} dynamic base patterns (template literals)`);
    }
    if (mergeablePatterns.size > 0) {
      console.log(`  \u2139 Found ${mergeablePatterns.size} mergeable patterns (className props)`);
    }
  }
  const patterns = config.include.length > 0 ? config.include : DEFAULT_PATTERNS;
  const allFiles = await findFiles(config.buildDir, patterns);
  const files = filterExcludedFiles(allFiles, config.exclude.files || []);
  if (config.verbose && allFiles.length !== files.length) {
    console.log(`  \u2139 Excluded ${allFiles.length - files.length} files from transformation`);
  }
  let filesModified = 0;
  let totalBytesChanged = 0;
  let totalSkippedForHydration = 0;
  const errors = [];
  for (const filePath of files) {
    if (!isJSFile(filePath) && !isHTMLFile(filePath) && !isRSCFile(filePath)) {
      continue;
    }
    const result = await transformFile(filePath, mappings, config, dryRun, dynamicBasePatterns, mergeablePatterns);
    if (result.error) {
      errors.push(result.error);
    }
    if (result.modified) {
      filesModified++;
      totalBytesChanged += result.replacements * 10;
    }
    totalSkippedForHydration += result.skippedForHydration;
  }
  if (config.verbose && totalSkippedForHydration > 0) {
    console.log(`Skipped ${totalSkippedForHydration} pattern transformations to prevent hydration mismatches`);
  }
  return {
    filesModified,
    bytesChanged: totalBytesChanged,
    errors
  };
}

export { transformBuildOutput };
